// @wanwu/base-vue-view-images/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('core-js/modules/es6.function.name');
require('core-js/modules/es6.array.is-array');
var _classCallCheck = _interopDefault(require('@babel/runtime-corejs2/helpers/esm/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime-corejs2/helpers/esm/createClass'));
require('core-js/modules/es6.number.constructor');
var baseVueSwipe = require('@wanwu/base-vue-swipe');
var popup = _interopDefault(require('@wanwu/base-vue-popup'));
var baseFn = require('@wanwu/base-fn');
var vueRuntimeHelpers = require('@wanwu/vue-runtime-helpers');
var Vue = _interopDefault(require('vue'));
var eventUtils = require('@wanwu/event-utils');

var script = {
  name: 'ViewImages',
  components: {
    Swipe: baseVueSwipe.Swipe,
    SwipeItem: baseVueSwipe.SwipeItem,
    popup: popup
  },
  props: {
    images: {
      type: Array
    },
    value: {
      type: Boolean,
      "default": false
    },
    currentImg: {
      type: Number
    },
    indicatorType: {
      // 可选值 'none','point','number'
      type: String,
      "default": 'point'
    },
    showSaveBtn: {
      type: Boolean
    },
    cut: {
      type: Boolean,
      "default": true
    },
    showH5SaveToast: {
      type: Boolean
    }
  },
  data: function data() {
    return {
      current: 0,
      clickIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/1470c5be5d081f53bf34fb3c435afdc0.png'
    };
  },
  computed: {},
  watch: {
    currentImg: function currentImg(val) {
      this.current = val;
    }
  },
  methods: {
    getRectangleImageWithWidth: baseFn.ImageUtils.getRectangleImageWithWidth,
    onSwipeChange: function onSwipeChange(index) {
      this.current = index;
    },
    popupClose: function popupClose() {
      this.$emit('input', false);
    },
    onInput: function onInput(val) {
      this.$emit('input', val);
    },
    saveImage: function saveImage() {
      this.$emit('saveImage', this.images[this.current], this.current, this.images);
    },
    onClosed: function onClosed() {
      this.$emit('closed');
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "component__viewImages viewImages"
  }, [_vm.images && _vm.images.length > 1 && _vm.indicatorType === 'number' && _vm.value ? _c('div', {
    staticClass: "number-indicator",
    attrs: {
      "slot": "external"
    },
    slot: "external"
  }, [_vm._v("\n        " + _vm._s(_vm.current + 1) + "/" + _vm._s(_vm.images.length) + "\n    ")]) : _vm._e(), _vm._v(" "), _c('popup', {
    staticClass: "popup",
    attrs: {
      "value": _vm.value,
      "overlay-style": {
        backgroundColor: '#000'
      },
      "slot-style": {
        height: '100%'
      }
    },
    on: {
      "input": _vm.onInput,
      "slotClick": _vm.popupClose,
      "closed": _vm.onClosed
    }
  }, [_c('Swipe', {
    staticClass: "swiper",
    attrs: {
      "show-indicators": false,
      "initial-swipe": _vm.currentImg,
      "autoplay": 0
    },
    on: {
      "change": _vm.onSwipeChange
    }
  }, _vm._l(_vm.images, function (item, index) {
    return _c('SwipeItem', {
      key: item + index,
      staticClass: "swiper__item flex f-ai-c f-jc-c"
    }, [_c('img', {
      directives: [{
        name: "lazy",
        rawName: "v-lazy",
        value: _vm.cut ? _vm.getRectangleImageWithWidth(item, 800) : item,
        expression: "cut ? getRectangleImageWithWidth(item, 800) : item"
      }],
      staticClass: "swiper__item--image"
    }), _vm._v(" "), _vm.showH5SaveToast ? _c('div', {
      staticClass: "popup__toast flex f-ai-c"
    }, [_c('img', {
      staticClass: "popup__toast--icon",
      attrs: {
        "src": _vm.clickIcon
      }
    }), _vm._v(" "), _c('span', {
      staticClass: "popup__toast--text"
    }, [_vm._v("长按保存或发送给朋友")])]) : _vm._e()]);
  }), 1)], 1), _vm._v(" "), _vm.images && _vm.images.length > 1 && _vm.indicatorType === 'point' && _vm.value ? _c('div', {
    staticClass: "point-indicator flex",
    attrs: {
      "slot": "external"
    },
    slot: "external"
  }, _vm._l(Array(_vm.images.length), function (item, index) {
    return _c('i', {
      key: index,
      "class": {
        point: true,
        active: _vm.current === index
      }
    });
  }), 0) : _vm._e(), _vm._v(" "), _vm.showSaveBtn && _vm.value ? _c('button', {
    staticClass: "popup__btn",
    attrs: {
      "slot": "external"
    },
    on: {
      "click": function click($event) {
        $event.stopPropagation();
        return _vm.saveImage($event);
      }
    },
    slot: "external"
  }, [_vm._v("\n        保存图片\n    ")]) : _vm._e()], 1);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-56f5413b_0", {
    source: ".component__viewImages .swiper[data-v-56f5413b]{width:7.5rem;height:100%}.component__viewImages .swiper__item[data-v-56f5413b]{overflow-y:auto}.component__viewImages .swiper__item--image[data-v-56f5413b]{width:7.5rem;display:block}.component__viewImages .number-indicator[data-v-56f5413b]{color:#fff;font-size:.26rem;display:inline-block;position:fixed;top:.5rem;left:50%;transform:translateX(-50%);z-index:4002}.component__viewImages .point-indicator[data-v-56f5413b]{position:fixed;bottom:.5rem;left:50%;transform:translateX(-50%);z-index:4002}.component__viewImages .point-indicator .point[data-v-56f5413b]{opacity:.3;border-radius:100%;width:.1rem;height:.1rem;transition:opacity .2s;background-color:#d8d8d8}.component__viewImages .point-indicator .point[data-v-56f5413b]:not(:last-child){margin-right:.1rem}.component__viewImages .point-indicator .active[data-v-56f5413b]{opacity:1;background-color:#ea313e}.component__viewImages[data-v-56f5413b] .popup-fade-enter,.component__viewImages[data-v-56f5413b] .popup-fade-leave-active{opacity:0;transform:translate3d(-50%,-50%,0)}.component__viewImages .popup__btn[data-v-56f5413b]{border:.01rem solid #fafafa;border-radius:.1rem;padding:.16rem .32rem;color:#fff;background-color:transparent;outline:0;position:fixed;bottom:.5rem;left:50%;transform:translateX(-50%);z-index:4002}.component__viewImages .popup__toast[data-v-56f5413b]{padding:.1rem .3rem;background-color:rgba(50,50,51,.88);color:#fff;font-size:.26rem;line-height:.4rem;border-radius:.16rem;position:absolute;top:50%;left:50%;transform:translate3d(-50%,-50%,0)}.component__viewImages .popup__toast--icon[data-v-56f5413b]{width:.5rem;display:block}.component__viewImages .popup__toast--text[data-v-56f5413b]{margin-left:.1rem;white-space:nowrap}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-56f5413b";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = vueRuntimeHelpers.normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, vueRuntimeHelpers.createInjector, undefined, undefined);

var ShowBigImgConstructor = Vue.extend(__vue_component__);
var instance;
var defaultProps = {
  images: null,
  currentImg: 0,
  indicatorType: 'point',
  showSaveBtn: false,
  cut: true,
  showH5SaveToast: false,
  removeFlag: true
};
var cacheEvents;

var ViewImg = /*#__PURE__*/function () {
  function ViewImg() {
    _classCallCheck(this, ViewImg);
  }

  _createClass(ViewImg, null, [{
    key: "show",

    /**
     *  查看大图
     * @param {string|props|{props,events}} options
     */
    value: function show(options) {
      var _this = this;

      if (!instance) {
        instance = new ShowBigImgConstructor().$mount();
        document.body.appendChild(instance.$el);
        instance.$on('input', function (val) {
          instance.value = val;
        });
        instance.$on('closed', function () {
          eventUtils.vmUnbindEvents(instance, cacheEvents);
          cacheEvents = null;

          if (instance.removeFlag) {
            _this.remove();
          }
        });
      }

      Object.assign(instance, defaultProps);

      if (options) {
        if (typeof options === 'string') {
          instance.images = [options];
          Object.assign(instance, {
            images: [options]
          });
        } else if (Array.isArray(options)) {
          // instance.images = options;
          Object.assign(instance, {
            images: options
          });
        } else if (options.props) {
          Object.assign(instance, options.props);
        } else {
          Object.assign(instance, options);
        }

        if (options.events) {
          cacheEvents = options.events;
          eventUtils.vmBindEvents(instance, cacheEvents);
        }
      }

      instance.value = true;
    }
    /**
     * 关闭查看大图弹窗
     */

  }, {
    key: "close",
    value: function close() {
      if (instance) {
        instance.value = false;
      }
    }
    /**
     * 删除dom节点，清除实例
     */

  }, {
    key: "remove",
    value: function remove() {
      if (instance) {
        var parent = instance.$el.parentNode;

        if (parent) {
          parent.removeChild(instance.$el);
        }

        instance.$destroy();
      }

      instance = null;
    }
  }]);

  return ViewImg;
}();

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
  Vue.prototype.$viewImg = ViewImg;
  Vue.viewImg = ViewImg;
};

exports.ViewImg = ViewImg;
exports.default = __vue_component__;
