// @wanwu/base-vue-overlay/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('core-js/modules/es6.function.name');
var _classCallCheck = _interopDefault(require('@babel/runtime-corejs2/helpers/esm/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime-corejs2/helpers/esm/createClass'));
var vueRuntimeHelpers = require('@wanwu/vue-runtime-helpers');
var Vue = _interopDefault(require('vue'));
var eventUtils = require('@wanwu/event-utils');

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script = {
  name: 'Overlay',
  components: {},
  props: {
    value: {
      type: Boolean,
      "default": false
    },
    overlayStyle: {
      type: [Object, String]
    },
    onClick: {
      type: Function
    }
  },
  data: function data() {
    return {};
  },
  computed: {},
  methods: {
    click: function click() {
      this.onClick && this.onClick();
      this.$emit('click');
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('transition', {
    attrs: {
      "name": "wanwu-fade"
    },
    on: {
      "afterEnter": function afterEnter($event) {
        return _vm.$emit('opened');
      },
      "afterLeave": function afterLeave($event) {
        return _vm.$emit('closed');
      }
    }
  }, [_vm.value ? _c('div', {
    staticClass: "component__overlay overlay",
    staticStyle: {
      "z-index": "3000"
    },
    style: _vm.overlayStyle,
    on: {
      "click": _vm.click
    }
  }) : _vm._e()]);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-4c4f3aee_0", {
    source: ".component__overlay[data-v-4c4f3aee]{position:fixed;top:0;left:0;width:100%;height:100%;background-color:rgba(0,0,0,.7)}@-webkit-keyframes wanwu-fade-in-data-v-4c4f3aee{from{opacity:0}to{opacity:1}}@keyframes wanwu-fade-in-data-v-4c4f3aee{from{opacity:0}to{opacity:1}}@-webkit-keyframes wanwu-fade-out-data-v-4c4f3aee{from{opacity:1}to{opacity:0}}@keyframes wanwu-fade-out-data-v-4c4f3aee{from{opacity:1}to{opacity:0}}.wanwu-fade-enter-active[data-v-4c4f3aee]{-webkit-animation:wanwu-fade-in-data-v-4c4f3aee .3s both ease;animation:wanwu-fade-in-data-v-4c4f3aee .3s both ease}.wanwu-fade-leave-active[data-v-4c4f3aee]{-webkit-animation:wanwu-fade-out-data-v-4c4f3aee .3s both ease;animation:wanwu-fade-out-data-v-4c4f3aee .3s both ease}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-4c4f3aee";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = vueRuntimeHelpers.normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, vueRuntimeHelpers.createInjector, undefined, undefined);

var OverlayConstructor = Vue.extend(__vue_component__);
var defaultOptions = {
  overlayStyle: null,
  onClick: null
};
var cacheEvents;

var Overlay = /*#__PURE__*/function () {
  function Overlay(options) {
    var _this = this;

    _classCallCheck(this, Overlay);

    var instance = new OverlayConstructor().$mount();
    document.body.appendChild(instance.$el);
    instance.$on('input', function (val) {
      instance.value = val;
    });
    instance.$on('closed', function () {
      eventUtils.vmUnbindEvents(_this.instance, cacheEvents);
      cacheEvents = null;

      _this.remove();
    });
    this.instance = instance;
    Object.assign(this.instance, defaultOptions);

    if (options instanceof Object) {
      if (options.props) {
        Object.assign(this.instance, options.props);
      } else {
        Object.assign(this.instance, options);
      }

      if (options.events) {
        cacheEvents = options.events;
        eventUtils.vmBindEvents(this.instance, cacheEvents);
      }
    } else if (options) {
      console.error('overlay参数仅支持对象，请传入正确的参数');
    }
  }

  _createClass(Overlay, [{
    key: "show",
    value: function show() {
      if (this.instance) {
        this.instance.value = true;
      }
    }
  }, {
    key: "close",
    value: function close() {
      if (this.instance) {
        this.instance.value = false;
      }
    }
  }, {
    key: "remove",
    value: function remove() {
      if (this.instance) {
        var parent = this.instance.$el.parentNode;

        if (parent) {
          parent.removeChild(this.instance.$el);
        }

        this.instance.$destroy();
      }
    }
  }]);

  return Overlay;
}();

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
  Vue.prototype.$overlay = Overlay;
  Vue.overlay = Overlay;
};

exports.Overlay = Overlay;
exports.default = __vue_component__;
