// @wanwu/base-vue-loading/lib/index.es.js
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.array.index-of';
import 'core-js/modules/es.number.constructor';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
// import 'iview';
var script = {
  name: 'Loading',
  props: {
    size: {
      type: Number,
      "default": 50
    },
    color: {
      type: String,
      "default": '#c9c9c9'
    },
    loadType: {
      type: String,
      "default": 'line',
      validator: function validator(value) {
        // 这个值必须匹配下列字符串中的一个
        return [// 环形线
        'line', // 点状
        'point'].indexOf(value) !== -1;
      }
    }
  },
  computed: {
    style: function style() {
      return {
        color: this.color,
        width: this.size / 100 + 'rem',
        height: this.size / 100 + 'rem'
      };
    },
    point: function point() {
      var point = [];
      var r = 19;
      var ox = 27;
      var oy = 27;
      var count = 8;
      var radians = Math.PI / 180 * Math.round(360 / count); // 弧度

      var i = 0;

      for (; i < count; i++) {
        var x = ox + r * Math.sin(radians * i);
        var y = oy + r * Math.cos(radians * i);
        point.unshift({
          x: x,
          y: y
        }); // 为保持数据顺时针
      }

      return point;
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "loading",
    style: _vm.style
  }, [_vm.loadType === 'line' ? _c('svg', {
    staticClass: "load",
    staticStyle: {
      "stroke": "currentColor",
      "fill": "none"
    },
    attrs: {
      "viewBox": "10 10 80 80"
    }
  }, [_c('circle', {
    staticClass: "circle",
    attrs: {
      "cx": "50",
      "cy": "50",
      "r": "30"
    }
  })]) : _c('svg', {
    staticClass: "load",
    staticStyle: {
      "fill": "currentColor",
      "stroke": "none"
    },
    attrs: {
      "x": "0px",
      "y": "0px",
      "viewBox": "0 0 54 54"
    }
  }, _vm._l(_vm.point, function (item, index) {
    return _c('circle', {
      key: index,
      attrs: {
        "cx": item.x,
        "cy": item.y,
        "r": 1.5 + 0.5 * index
      }
    });
  }), 0)]);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-e2757bee_0", {
    source: ".loading[data-v-e2757bee]{display:inline-block}.div[data-v-e2757bee]{display:flex}.load[data-v-e2757bee]{-webkit-animation:rotate-data-v-e2757bee 2s linear infinite;animation:rotate-data-v-e2757bee 2s linear infinite;width:100%}.circle[data-v-e2757bee]{stroke-width:3;-webkit-animation:dash-data-v-e2757bee 1.5s linear infinite;animation:dash-data-v-e2757bee 1.5s linear infinite}@-webkit-keyframes rotate-data-v-e2757bee{0%{transform:rotate(0)}100%{transform:rotate(360deg)}}@keyframes rotate-data-v-e2757bee{0%{transform:rotate(0)}100%{transform:rotate(360deg)}}@-webkit-keyframes dash-data-v-e2757bee{0%{stroke-dasharray:1,200;stroke-dashoffset:0}50%{stroke-dasharray:130,200;stroke-dashoffset:-50}100%{stroke-dasharray:130,200;stroke-dashoffset:-188}}@keyframes dash-data-v-e2757bee{0%{stroke-dasharray:1,200;stroke-dashoffset:0}50%{stroke-dasharray:130,200;stroke-dashoffset:-50}100%{stroke-dasharray:130,200;stroke-dashoffset:-188}}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-e2757bee";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
