// @wanwu/base-vue-tabs/src/mixins/props.js
export default {
    props: {
        value: [
            String,
            Number,
        ],
        shouldAnimate: {
            type: Boolean,
            default: true,
        },
        duration: {
            type: Number,
            default: 0.3,
        },
        sticky: {
            type: Boolean,
            default: false,
        },
        tabHeight: {
            type: Number,
            default: 0,
        },
        // 单位rem（fixed定位时）
        top: {
            type: Number,
            default: 0,
        },
        // 滚动阈值，标签数量超过阈值时开始横向滚动
        swipeThreshold: {
            type: [
                Number,
                String,
            ],
            default: 4,
        },
        // 滑动切换标签页
        swipeable: {
            type: Boolean,
            default: false,
        },
        // 内容切换动画
        animated: {
            type: Boolean,
            default: false,
        },
        // 样式风格类型，可选值为card,icon
        type: {
            type: String,
            default: 'line',
        },
        // 标签栏样式
        customTabbarStyle: {
            type: Object,
            default: () => ({}),
        },
        // 标签样式
        customTabStyle: {
            type: Object,
            default: () => ({}),
        },
        // 选中样式
        customActiveStyle: {
            type: Object,
            default: () => ({}),
        },
        // 下划线样式
        customLineStyle: {
            type: Object,
            default: () => ({}),
        },
        // 是否自动填充吸顶时content的高度
        fillScreen: {
            type: Boolean,
            default: false,
        },
        // 距离页面指定到的距离时候吸顶
        offsetTop: {
            type: Number,
            default: 0,
        },
        // 是否要开启延迟切换
        delay: {
            type: Boolean,
            default: false,
        },
        // icon类型的图片样式
        iconStyle: {
            type: Object,
            default: () => ({}),
        },
        showAll: {
            type: Boolean,
            default: false,
        },
        showAllStyle: {
            type: Object,
            default: () => ({}),
        },
        fixed: {
            type: Boolean,
            default: false,
        },
        // 标签宽度自适应
        tabAutoWidth: {
            type: Boolean,
            default: false,
        },
        // type === 'newModel' 使用默认下划线插槽
        useSlotLine: {
            type: Boolean,
            default: true,
        },
        // 滚动导航
        scrollspy: {
            type: Boolean,
            default: false,
        },
        // 隐藏bar
        hiddenBar: {
            type: [
                Boolean,
                String,
            ],
            default: '',
        },
        // 是否开启滚动导航动画
        scrollspyAnimate: {
            type: Boolean,
            default: true,
        },
        // 保持吸顶，不跟随内容模块移出视图
        isStickyAlways: {
            type: Boolean,
            default: false,
        },
        // 吸顶状态消失时标签栏动画
        tabbarAnimate: {
            type: Boolean,
            default: true,
        },
        splitLine: {
            type: [Boolean, undefined],
            default: undefined,
        },
        // tab切换时滚动条是否自动滚动到顶部
        scrollToTopOnChange: {
            type: Boolean,
            default: true,
        },
        // change前调用
        // @return {Promise | Boolean}
        beforeChange: {
            type: Function,
        },
    },
};
