// @wanwu/base-request/src/common/tokenManager.js
import { fn, ua } from '@wanwu/base-fn';
import Config from '@wanwu/mall-web-config';
export const saveToken = function (token) {
    fn.setCookie(Config.CookieTokenKey, token, {
        path: '/',
        domain: '.wanwudezhi.com',
        expires: 7,
    });
    localStorage.setItem(Config.CookieTokenKey, token);
};

export const removeToken = function () {
    localStorage.removeItem(Config.CookieTokenKey);
    fn.removeCookie(Config.CookieTokenKey, {
        path: '/',
        domain: '.wanwudezhi.com',
    });
};

export const getToken = function (getTokenOnUrl) {
    // 获取用户token值
    let token;
    if (ua.isApp() || ua.isWechatMiniProgram() || getTokenOnUrl) {
        token = fn.getQueryString('token') || fn.getCookie(Config.CookieTokenKey) || localStorage.getItem(Config.CookieTokenKey);
        if (token && token.indexOf('Wwdz (null)') !== -1) {
            token = '';
        }
    } else {
        token = fn.getCookie(Config.CookieTokenKey) || localStorage.getItem(Config.CookieTokenKey);
    }
    return token;
};
