// @wanwu/get-wx-qr/src/index.js
import Share from '@wanwu/base-share';
import axios from '@wanwu/base-request';
import { fn,ua } from '@wanwu/base-fn';
import User from '@wanwu/base-user';
import * as QRCode from 'qrcode';
import {
    activeGoodsType,
} from '@wanwu/resource-center';
import {
    getShortChain,
} from '@wanwu/base-vue-share-dialog';


// 生成小程序二维码
export const getXcxCode = async (shareOptions, skipLogin, pushTarget = 'xcx') => {
    // const shareOptions = Share.getCacheShareOptions();
    let programAppId;
    try {
        const res = await activeGoodsType();
        programAppId = res.programAppId;
    } catch {
        programAppId = 'wx98b7b078cfb7164d';
    }
    const shareType = pushTarget === 'xcx';
    shareOptions = Share.getCacheShareOptions(shareOptions);
    const inviteCode = User.getUserInfo().inviteCode || fn.getQueryString('inviteCode');
    // return axios.post('https://api.wanwudezhi.com/media/xcx/getWxCode', {
    //     appId: 'wxb2de1255783f7c70',
    return axios.post('https://api.wanwudezhi.com/media/xcx/getXcxCode', {
        appId: programAppId || 'wx98b7b078cfb7164d',
        image: shareOptions.imgUrl,
        page: shareType ? ('pages' + shareOptions.page) : 'pages/webview/index',
        title: shareOptions.title,
        desc: shareOptions.desc && shareOptions.desc.substr(0, 50),
        url: (shareType ? shareOptions.queryStr : `src=${encodeURIComponent(shareOptions.link)}&inviteCode=${inviteCode}`) || '',
    }, {
        skipLogin,
    }).then(res => {
        return res;
    }).catch(err => {
        const QRUrl = 'https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png';
        if (skipLogin) {
            return QRUrl;
        } else {
            return Promise.reject(err);
        }
    });
};

// 获取小程序码，暴露设置样式的style参数
export async function getQrCodeByStyle({
    shareOptions, skipLogin, pushTarget = 'xcx', style = {},
}) {
    const shareType = pushTarget === 'xcx';
    shareOptions = Share.getCacheShareOptions(shareOptions);
    const inviteCode = User.getUserInfo().inviteCode || fn.getQueryString('inviteCode');
    let wxAppId;
    try {
        const res = await activeGoodsType();
        wxAppId = res.wxAppId;
    } catch {
        wxAppId = 'wxb2de1255783f7c70';
    }
    return axios.post('https://api.wanwudezhi.com/media/xcx/getWxCodeLink', {
        appId: wxAppId || 'wxb2de1255783f7c70',
        image: shareOptions.imgUrl,
        page: shareType ? ('pages' + shareOptions.page) : 'pages/webview/index',
        title: shareOptions.title,
        desc: shareOptions.desc && shareOptions.desc.substr(0, 50),
        url: shareType ? shareOptions.queryStr : `src=${encodeURIComponent(shareOptions.link)}&inviteCode=${inviteCode}`,
    }, {
        skipLogin,
    }).then((link) => {
        return QRCode.toDataURL(link, style || {});
    }).catch(err => {
        const QRUrl = 'https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png';
        if (skipLogin) {
            return QRUrl;
        } else {
            return Promise.reject(err);
        }
    });
}

// 生成h5二维码
export async function getH5Code(shareOptions, style) {
    if (ua.isWechatMiniProgram()) {
        // 小程序中使用H5,分享使用小程序的码
        const domain = window.location.href;
        return domain &&
         (await axios.post("https://api.wanwudezhi.com/media/xcx/getXcxCode", {
            appId: "wx98b7b078cfb7164d",
            page: "pages/webview/index",
            url: `src=${encodeURI(domain)}`,
            width: 156
        },{
            skipLogin:true
        }).then((res) => {
            return res;
        }).catch(err => {
            const QRUrl = 'https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png';
            return QRUrl;
        }));
    }else{
        const shareOpt = Share.getCacheShareOptions(shareOptions);
        const link = (shareOpt && shareOpt.link) ? shareOpt.link : location.href;
        await User.ensureLogin();
        let shortLink = link;
        try {
            shortLink = await getShortChain(shareOpt);
            return QRCode.toDataURL(shortLink, style || {});
        } catch (error) {
            const QRUrl = "https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png";
            return QRUrl;
        }
    }
}


/**
 * 支持配置小程序码或二维码图片
 *
 * @param {object} options
 * @param {object} options.shareOptions 与当前页面分享配置不同的分享配置对象
 * @param {boolean} options.skipLogin 是否强制登陆
 * @param {string} options.pushTarget 分享小程序原生路径还是小程序webview h5路径 ['xcx','h5']
 * @param {number} options.codeType code码类型 【0:h5二维码,1,小程序码,2:公众号二维码】
 * @param {object} options.style 二维码样式配置 文档查询  https://www.npmjs.com/package/qrcode
 * @returns {Promise} 返回promise，resolve 二维码或者小程序码的链接或者base-64
 */
export async function getShareCodeImg(options = {}) {
    // options.shareOptions = Share.getCacheShareOptions(options.shareOptions);
    const {
        shareOptions, skipLogin, pushTarget = 'xcx', style = {}, codeType = 0,
    } = options;
    if (codeType === 2) {
        return getQrCodeByStyle(options);
    } else if (codeType === 0) {
        return getH5Code(shareOptions, style);
    } else {
        return getXcxCode(shareOptions, skipLogin, pushTarget);
    }
}
export default getH5Code;
