// @wanwu/base-logger/src/config.ts
import {ConfigTypes} from "./util/types";
import hash from "./util/hash";
import { ua } from "@wanwu/base-fn";

const priceTrace = `${ua.isWechatMiniProgram() ? 'XCX' : 'H5'}.${hash.rand(8)}`;

const config: ConfigTypes = {
    LogUrl: 'https://api.wanwudezhi.com/api/logcenter/batchLog', // 打点地址
    NewLogUrl: 'https://api.wanwudezhi.com/api/logcenter/postCommonUelLog?logsTore=wwdz_action_log', // 新版打点地址
    CORS__LogUrl: '/log/api/logcenter/batchLog', // cors 打点接口地址
    NEW__CORS__LogUrl: '/log/api/logcenter/postCommonUelLog', // cors 打点接口地址
    // NEW__CORS__LogUrl: '/log/api/logcenter/batchUploadLog', // 最新的埋点接口地址
    UserInfoKey: 'userInfo', // 获取用户信息的字段
    rtp_cnt_a: '', // 指定rtp_cnt的a字段
    rtp_cnt_b: '',  // 指定rtp_cnt的b字段
    eventId: 10001,  // 默认PV eventId
    platform: '',  // 平台，后期可拓展方法进行平台获取
    ignorePageRegExp: [], // 不需要上报的页面路由(无任何打点事件)
    send_interval: 2000,  // 间隔多长时间发送日志
    one_send_max_length: 4, // 一次请求最大发送几条数据，防止数据太大
    filters: [], // 当前页面是否包含过滤器
    env_production_host: /^h5\.wanwudezhi\.com/, // 正式环境的域名匹配
    // 业务追踪起止字段
    traceLink: {
        priceTrace
    }
};

export default config;
