// @wanwu/base-vue-loading-page/src/index.js
import comp from './index.vue';
import Vue from 'vue';
import getCompDefaultOptions from '@wanwu/get-comp-default-options';
const LoadingPageConstructor = Vue.extend(comp);
let instance;
let _closeFn;
const closeMixin = {
    destroyed() {
        _closeFn && _closeFn();
    },
    deactivated() {
        _closeFn && _closeFn();
    },
};
const defaultOptions = getCompDefaultOptions(comp);
//  {
//     size: 100,
//     loadingText: '',
//     bgColor: '#f7f7f7',
// };
class LoadingPage {
    static show(options) {
        if (!instance) {
            instance = new LoadingPageConstructor().$mount(document.createElement('div'));
            document.body.appendChild(instance.$el);
            instance.$on('input', (val) => {
                instance.value = val;
            });
        }
        Object.assign(instance, defaultOptions);
        if (options) {
            if (typeof options === 'string') {
            // instance.loadingText = options;
                Object.assign(instance, {
                    loadingText: options,
                });
            } else {
                Object.assign(instance, options);
            }
        }
        instance.value = true;
    }

    static close() {
        if (instance && instance.value) {
            instance.value = false;
        }
        _closeFn && _closeFn();
    }
}
comp.install = (Vue, {
    closeFn,
}) => {
    comp.mixin = comp.mixin || [];
    comp.mixin.push(closeMixin);
    Vue.component(comp.name, comp);
    Vue.prototype.$loadingPage = LoadingPage;
    Vue.loadingPage = LoadingPage;
    _closeFn = closeFn;
};
export {
    LoadingPage,
};
export default comp;
