// @wanwu/base-fn/lib.es/image/ImageUtils.js
/**
 * Create by changsheng on 2019-04-03 11:37
 */
import { typeYalidator } from '../fn/index';
let isSupportWebP;
const cdnUrlPattern = /^http(s?):\/\/cdn\.wanwudezhi\.com/;
/**
 * 判断是否支持webp
 * @return {boolean}
 */
export function checkSupportWebP() {
    if (isSupportWebP) {
        return isSupportWebP;
    }
    try {
        isSupportWebP = (document.createElement('canvas').toDataURL('image/webp').indexOf('data:image/webp') == 0);
        return isSupportWebP;
    }
    catch (err) {
        return false;
    }
}
// export const isSupportWebP = checkSupportWebP();
/**
 * 取400*400的方图
 * @param url
 * @param {string} format
 * @return {string}
 */
export function getSmallImage(url, format) {
    return getSquareImage(url, 400, format);
}
/**
 * 取变成一样的图
 * @param url 图片地址
 * @param length 边长
 * @param {string} format 默认转成jpg
 * @return {string}
 */
export function getSquareImage(url, length, format) {
    return getRectangleImage(url, { width: length, height: length, cut: true, format });
}
/**
 * 取矩形图，需要传长宽
 * @param url
 * @param {{cut?: boolean; width?: number; height?: number; format?: string}} options
 * @return {string}
 */
export function getRectangleImage(url, options) {
    if (!url) {
        return url;
    }
    if (!cdnUrlPattern.test(url)) {
        return url;
    }
    if (!options) {
        options = {};
    }
    const originUrl = getOriginImage(url);
    let mode = '2';
    if (options.cut) {
        mode = '1';
    }
    // 获取文件默认后缀
    const suffix = originUrl.replace(/^.*\.(\w+)$/, '$1');
    // 导出最终的图片链接
    const videoReg = /(mp4|rm(vb)?|flv|mp(e)?(g)?|(n)?avi|wmv|m3u8|mov|mtv|asf|dat|mov|3gp|amv|dmv|m2v|vob|f4v|vcd)/;
    const offsetReg = /^.*\/offset\/(\w+)/;
    // format
    const format = checkSupportWebP() ? 'webp' : (options.format || (videoReg.test(suffix.toLowerCase()) ? 'png' : suffix) || 'jpg');
    // imageView2Src后缀
    let imageView2Src;
    if (options.width && options.height) {
        imageView2Src = `imageView2/${mode}/w/${options.width}/h/${options.height}/format/${format}`;
    }
    else if (options.width) {
        imageView2Src = `imageView2/${mode}/w/${options.width}/format/${format}`;
    }
    else if (options.height) {
        imageView2Src = `imageView2/${mode}/h/${options.height}/format/${format}`;
    }
    if (videoReg.test(suffix.toLowerCase())) {
        // 视频图片
        let _offset;
        if (typeYalidator.isEmpty(options.offset)) {
            if (offsetReg.test(url)) {
                _offset = url.replace(/^.*\/offset\/(\w+)/, '$1');
            }
            else {
                _offset = 0;
            }
        }
        else {
            _offset = options.offset;
        }
        return `${originUrl}?vframe/png/offset/${_offset}|${imageView2Src}`;
    }
    else {
        // 普通cdn图片
        return `${originUrl}?${imageView2Src}`;
    }
}
/**
 * 取矩形图，制定宽度，不裁剪
 * @param url 图片url
 * @param width 宽度
 * @param {string} format 后缀
 * @return {string}
 */
export function getRectangleImageWithWidth(url, width, format) {
    return getRectangleImage(url, { width, format });
}
/**
 * 取矩形图，制定宽度，不裁剪
 * @param url 图片url
 * @param height 高度
 * @param {string} format 后缀
 * @return {string}
 */
export function getRectangleImageWithHeight(url, height, format) {
    return getRectangleImage(url, { height, format });
}
export function getOriginImage(url, watermark = false) {
    if (!checkIsCdnUrl(url)) {
        return url;
    }
    return url && url.split('?')[0] || '';
}
export function getImageWithWatermark(url, watermarkString) {
    if (!checkIsCdnUrl(url)) {
        return url;
    }
    const originUrl = getOriginImage(url);
    if (originUrl) {
        const watermark = watermarkString || 'watermark/1/image/aHR0cDovL3BvcnRhbC1vcmlnaW4ud2Fud3VkZXpoaS5jb20vd2F0ZXJtYXJrX3YxLnBuZw==/dissolve/100/gravity/SouthEast/dx/10/dy/10';
        return `${originUrl}?${watermark}`;
    }
    return originUrl;
}
export function checkIsCdnUrl(url) {
    return cdnUrlPattern.test(url);
}
