// @wanwu/base-vue-empty/lib/index.es.js
import 'core-js/modules/es.function.name';
import Loading from '@wanwu/base-vue-loading';
import '@wanwu/base-style';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';

//
var script = {
  name: 'Empty',
  components: {
    Loading: Loading
  },
  props: {
    desc: {
      type: String
    },
    buttonText: {
      type: String
    },
    onButtonClick: {
      type: Function
    },
    loadingStatus: {
      type: String
    },
    image: {
      type: String,
      "default": 'https://cdn.wanwudezhi.com/seller-admin/image/MTU1MzA3ODUzOTM4Mw==.png'
    }
  },
  data: function data() {
    return {};
  },
  computed: {},
  methods: {}
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "component__empty empty flex f-ff-cn f-ai-c f-jc-c"
  }, [_vm.loadingStatus !== 'loading' ? _c('img', {
    staticClass: "image",
    attrs: {
      "src": _vm.image
    }
  }) : _vm._e(), _vm._v(" "), _vm.loadingStatus === 'loading' ? _c('div', {
    staticClass: "desc"
  }, [_c('Loading', {
    staticClass: "Loading",
    attrs: {
      "size": 80
    }
  })], 1) : _vm.desc ? _c('div', {
    staticClass: "desc"
  }, [_vm._v("\n        " + _vm._s(_vm.desc) + "\n    ")]) : _vm._e(), _vm._v(" "), _vm.buttonText ? _c('button', {
    staticClass: "button btn",
    on: {
      "click": _vm.onButtonClick
    }
  }, [_vm._v("\n        " + _vm._s(_vm.buttonText) + "\n    ")]) : _vm._e()]);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-768ee2a4_0", {
    source: ".component__empty[data-v-768ee2a4]{height:10rem}.component__empty .image[data-v-768ee2a4]{width:2.88rem;height:2.56rem}.component__empty .desc[data-v-768ee2a4]{margin-top:.3rem;font-size:.28rem;font-weight:400;color:#b4b4b4;line-height:.4rem;padding:0 1rem;max-width:7.5rem}.component__empty .button[data-v-768ee2a4]{width:5.7rem;height:.9rem;background:#ea3131;border-radius:.45rem;font-size:.3rem;font-weight:500;color:#fff;line-height:.42rem;margin-top:2.5rem}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-768ee2a4";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
