// @wanwu/base-vue-bottom-x/lib/index.es.js
import 'core-js/modules/es.function.name';
import { ua } from '@wanwu/base-fn';
import occupying from '@wanwu/base-vue-occupying';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';

//
var script = {
  name: 'BottomX',
  components: {
    occupying: occupying
  },
  props: {
    occupyingColor: {
      type: String,
      "default": '#FFFFFF'
    }
  },
  data: function data() {
    return {
      isIPhoneX: false
    };
  },
  created: function created() {
    this.isIPhoneX = ua.isIPhoneX();
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "component__bottomX bottomX"
  }, [_vm._t("default"), _vm._v(" "), _vm.isIPhoneX ? _c('occupying', {
    attrs: {
      "height": 56,
      "bgColor": _vm.occupyingColor
    }
  }) : _vm._e()], 2);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-51e6fd85_0", {
    source: ".bottomX[data-v-51e6fd85]{position:fixed;bottom:0;left:50%;transform:translate(-50%,0);z-index:2000}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-51e6fd85";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
