// @wanwu/base-vue-bottom-x/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('core-js/modules/es.function.name');
var baseFn = require('@wanwu/base-fn');
var occupying = _interopDefault(require('@wanwu/base-vue-occupying'));
var vueRuntimeHelpers = require('@wanwu/vue-runtime-helpers');

//
var script = {
  name: 'BottomX',
  components: {
    occupying: occupying
  },
  props: {
    occupyingColor: {
      type: String,
      "default": '#FFFFFF'
    }
  },
  data: function data() {
    return {
      isIPhoneX: false
    };
  },
  created: function created() {
    this.isIPhoneX = baseFn.ua.isIPhoneX();
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "component__bottomX bottomX"
  }, [_vm._t("default"), _vm._v(" "), _vm.isIPhoneX ? _c('occupying', {
    attrs: {
      "height": 56,
      "bgColor": _vm.occupyingColor
    }
  }) : _vm._e()], 2);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-51e6fd85_0", {
    source: ".bottomX[data-v-51e6fd85]{position:fixed;bottom:0;left:50%;transform:translate(-50%,0);z-index:2000}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-51e6fd85";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = vueRuntimeHelpers.normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, vueRuntimeHelpers.createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

exports.default = __vue_component__;
