// @wanwu/base-vue-scroll/lib/index.es.js
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.array.for-each';
import 'core-js/modules/es.number.constructor';
import 'core-js/modules/web.dom-collections.for-each';
import loading from '@wanwu/base-vue-loading';
import { fn } from '@wanwu/base-fn';
import '@wanwu/base-style';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';

var throttle = fn.throttle;
var script = {
  name: 'Scroll',
  components: {
    loading: loading
  },
  props: {
    loadingText: {
      type: String,
      "default": '加载中...'
    },
    noMoreText: {
      type: String,
      "default": '没有更多了'
    },
    loadingStyle: {
      type: [Object, String]
    },
    hasMore: {
      type: Boolean,
      "default": true
    },
    distanceToEdgeOffset: {
      type: Number,
      "default": 2
    },
    onReachBottom: {
      type: Function
    },
    slotClass: {
      type: [Object, String]
    },
    loading: {
      type: Boolean,
      "default": false
    },
    showText: {
      type: Boolean,
      "default": true
    }
  },
  data: function data() {
    return {
      handleScroll: function handleScroll() {},
      atBottom: false,
      events: ['scroll', 'wheel', 'mousewheel', 'resize', 'animationend', 'transitionend', 'touchmove']
    };
  },
  created: function created() {
    this.handleScroll = throttle(this.scrollActive, 150); // window.addEventListener('scroll', this.handleScroll);

    this.bindEvents(true);
  },
  mounted: function mounted() {
    this.$nextTick(this.handleScroll);
  },
  activated: function activated() {
    // window.addEventListener('scroll', this.handleScroll);
    this.bindEvents(true);
  },
  deactivated: function deactivated() {
    // window.removeEventListener('scroll', this.handleScroll);
    this.bindEvents(false);
  },
  beforeDestroy: function beforeDestroy() {
    // window.removeEventListener('scroll', this.handleScroll);
    this.bindEvents(false);
  },
  methods: {
    scrollActive: function scrollActive() {
      // if (this.hasMore) {
      //     let bottomY = this.$refs.bottom.getBoundingClientRect().y;
      //     let containerBottom =
      //         this.$refs.slot__container.getBoundingClientRect().y +
      //         this.$refs.slot__container.getBoundingClientRect().height;
      //     let toEdge = bottomY - containerBottom;
      //     if (toEdge < this.distanceToEdge && !this.loading) {
      //         this.onReachBottom && this.onReachBottom();
      //     }
      // }
      if (this.hasMore && !this.loading) {
        var docE = document.documentElement;
        var scrollTop = window.scrollY;
        var clientHeight = docE.clientHeight; // const bottomY = this.$refs.bottom.getBoundingClientRect().top;

        var scrollHeight = docE.scrollHeight;

        if (scrollTop + clientHeight >= scrollHeight - this.distanceToEdgeOffset * clientHeight) {
          if (this.onReachBottom) {
            this.onReachBottom();
          }

          this.$emit('onReachBottom'); // this.onReachBottom && this.onReachBottom();
        }
      }
    },
    bindEvents: function bindEvents(isBind) {
      var _this = this;

      this.events.forEach(function (event) {
        window[isBind ? 'addEventListener' : 'removeEventListener'](event, _this.handleScroll);
      });
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "components__scroll"
  }, [_c('div', {
    staticClass: "scroll__container"
  }, [_c('div', {
    ref: "slot__container",
    staticClass: "slot__container",
    "class": _vm.slotClass
  }, [_vm._t("default"), _vm._v(" "), _c('transition', {
    attrs: {
      "name": "wanwu-slide-up"
    }
  }, [_vm._t("bottom", [_vm.showText ? _c('div', {
    ref: "bottom",
    staticClass: "bottom flex f-ai-c f-jc-c",
    style: _vm.loadingStyle
  }, [_vm.hasMore ? _c('loading', {
    staticClass: "loading",
    attrs: {
      "size": 26
    }
  }) : _vm._e(), _vm._v("\n                        " + _vm._s(_vm.hasMore ? _vm.loadingText : _vm.noMoreText) + "\n                    ")], 1) : _vm._e()])], 2)], 2)])]);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-470dd6d7_0", {
    source: ".slot__container .bottom[data-v-470dd6d7]{width:100%;overflow:hidden;font-size:.26rem;text-align:center;height:.8rem;color:#c9c9c9}.slot__container .bottom .loading[data-v-470dd6d7]{margin-right:.15rem}@-webkit-keyframes wanwu-slide-up-enter-data-v-470dd6d7{from{transform:translate3d(0,100%,0)}}@keyframes wanwu-slide-up-enter-data-v-470dd6d7{from{transform:translate3d(0,100%,0)}}@-webkit-keyframes wanwu-slide-up-leave-data-v-470dd6d7{to{transform:translate3d(0,100%,0)}}@keyframes wanwu-slide-up-leave-data-v-470dd6d7{to{transform:translate3d(0,100%,0)}}.wanwu-slide-up-enter-active[data-v-470dd6d7]{-webkit-animation:wanwu-slide-up-enter-data-v-470dd6d7 .3s both ease;animation:wanwu-slide-up-enter-data-v-470dd6d7 .3s both ease}.wanwu-slide-up-leave-active[data-v-470dd6d7]{-webkit-animation:wanwu-slide-up-leave-data-v-470dd6d7 .3s both ease;animation:wanwu-slide-up-leave-data-v-470dd6d7 .3s both ease}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-470dd6d7";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
