// @wanwu/get-wx-qr/lib/index.es.js
import 'core-js/modules/es.array.concat';
import 'core-js/modules/es.object.to-string';
import 'core-js/modules/es.promise';
import 'core-js/modules/es.string.link';
import 'regenerator-runtime/runtime';
import _asyncToGenerator from '@babel/runtime/helpers/esm/asyncToGenerator';
import Share from '@wanwu/base-share';
import axios from '@wanwu/base-request';
import { fn, ua } from '@wanwu/base-fn';
import User from '@wanwu/base-user';
import { toDataURL } from 'qrcode';
import { activeGoodsType } from '@wanwu/resource-center';
import { getShortChain } from '@wanwu/base-vue-share-dialog';

var getXcxCode = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(shareOptions, skipLogin) {
    var pushTarget,
        programAppId,
        res,
        shareType,
        inviteCode,
        _args = arguments;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            pushTarget = _args.length > 2 && _args[2] !== undefined ? _args[2] : 'xcx';
            _context.prev = 1;
            _context.next = 4;
            return activeGoodsType();

          case 4:
            res = _context.sent;
            programAppId = res.programAppId;
            _context.next = 11;
            break;

          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](1);
            programAppId = 'wx98b7b078cfb7164d';

          case 11:
            shareType = pushTarget === 'xcx';
            shareOptions = Share.getCacheShareOptions(shareOptions);
            inviteCode = User.getUserInfo().inviteCode || fn.getQueryString('inviteCode'); // return axios.post('https://api.wanwudezhi.com/media/xcx/getWxCode', {
            //     appId: 'wxb2de1255783f7c70',

            return _context.abrupt("return", axios.post('https://api.wanwudezhi.com/media/xcx/getXcxCode', {
              appId: programAppId || 'wx98b7b078cfb7164d',
              image: shareOptions.imgUrl,
              page: shareType ? 'pages' + shareOptions.page : 'pages/webview/index',
              title: shareOptions.title,
              desc: shareOptions.desc && shareOptions.desc.substr(0, 50),
              url: (shareType ? shareOptions.queryStr : "src=".concat(encodeURIComponent(shareOptions.link), "&inviteCode=").concat(inviteCode)) || ''
            }, {
              skipLogin: skipLogin
            }).then(function (res) {
              return res;
            })["catch"](function (err) {
              var QRUrl = 'https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png';

              if (skipLogin) {
                return QRUrl;
              } else {
                return Promise.reject(err);
              }
            }));

          case 15:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[1, 8]]);
  }));

  return function getXcxCode(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}(); // 获取小程序码，暴露设置样式的style参数

function getQrCodeByStyle(_x3) {
  return _getQrCodeByStyle.apply(this, arguments);
} // 生成h5二维码

function _getQrCodeByStyle() {
  _getQrCodeByStyle = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(_ref2) {
    var shareOptions, skipLogin, _ref2$pushTarget, pushTarget, _ref2$style, style, shareType, inviteCode, wxAppId, res;

    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            shareOptions = _ref2.shareOptions, skipLogin = _ref2.skipLogin, _ref2$pushTarget = _ref2.pushTarget, pushTarget = _ref2$pushTarget === void 0 ? 'xcx' : _ref2$pushTarget, _ref2$style = _ref2.style, style = _ref2$style === void 0 ? {} : _ref2$style;
            shareType = pushTarget === 'xcx';
            shareOptions = Share.getCacheShareOptions(shareOptions);
            inviteCode = User.getUserInfo().inviteCode || fn.getQueryString('inviteCode');
            _context2.prev = 4;
            _context2.next = 7;
            return activeGoodsType();

          case 7:
            res = _context2.sent;
            wxAppId = res.wxAppId;
            _context2.next = 14;
            break;

          case 11:
            _context2.prev = 11;
            _context2.t0 = _context2["catch"](4);
            wxAppId = 'wxb2de1255783f7c70';

          case 14:
            return _context2.abrupt("return", axios.post('https://api.wanwudezhi.com/media/xcx/getWxCodeLink', {
              appId: wxAppId || 'wxb2de1255783f7c70',
              image: shareOptions.imgUrl,
              page: shareType ? 'pages' + shareOptions.page : 'pages/webview/index',
              title: shareOptions.title,
              desc: shareOptions.desc && shareOptions.desc.substr(0, 50),
              url: shareType ? shareOptions.queryStr : "src=".concat(encodeURIComponent(shareOptions.link), "&inviteCode=").concat(inviteCode)
            }, {
              skipLogin: skipLogin
            }).then(function (link) {
              return toDataURL(link, style || {});
            })["catch"](function (err) {
              var QRUrl = 'https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png';

              if (skipLogin) {
                return QRUrl;
              } else {
                return Promise.reject(err);
              }
            }));

          case 15:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2, null, [[4, 11]]);
  }));
  return _getQrCodeByStyle.apply(this, arguments);
}

function getH5Code(_x4, _x5) {
  return _getH5Code.apply(this, arguments);
}
/**
 * 支持配置小程序码或二维码图片
 *
 * @param {object} options
 * @param {object} options.shareOptions 与当前页面分享配置不同的分享配置对象
 * @param {boolean} options.skipLogin 是否强制登陆
 * @param {string} options.pushTarget 分享小程序原生路径还是小程序webview h5路径 ['xcx','h5']
 * @param {number} options.codeType code码类型 【0:h5二维码,1,小程序码,2:公众号二维码】
 * @param {object} options.style 二维码样式配置 文档查询  https://www.npmjs.com/package/qrcode
 * @returns {Promise} 返回promise，resolve 二维码或者小程序码的链接或者base-64
 */

function _getH5Code() {
  _getH5Code = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(shareOptions, style) {
    var domain, shareOpt, link, shortLink, QRUrl;
    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            if (!ua.isWechatMiniProgram()) {
              _context3.next = 10;
              break;
            }

            // 小程序中使用H5,分享使用小程序的码
            domain = window.location.href;
            _context3.t0 = domain;

            if (!_context3.t0) {
              _context3.next = 7;
              break;
            }

            _context3.next = 6;
            return axios.post("https://api.wanwudezhi.com/media/xcx/getXcxCode", {
              appId: "wx98b7b078cfb7164d",
              page: "pages/webview/index",
              url: "src=".concat(encodeURI(domain)),
              width: 156
            }, {
              skipLogin: true
            }).then(function (res) {
              return res;
            })["catch"](function (err) {
              var QRUrl = 'https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png';
              return QRUrl;
            });

          case 6:
            _context3.t0 = _context3.sent;

          case 7:
            return _context3.abrupt("return", _context3.t0);

          case 10:
            shareOpt = Share.getCacheShareOptions(shareOptions);
            link = shareOpt && shareOpt.link ? shareOpt.link : location.href;
            _context3.next = 14;
            return User.ensureLogin();

          case 14:
            shortLink = link;
            _context3.prev = 15;
            _context3.next = 18;
            return getShortChain(shareOpt);

          case 18:
            shortLink = _context3.sent;
            return _context3.abrupt("return", toDataURL(shortLink, style || {}));

          case 22:
            _context3.prev = 22;
            _context3.t1 = _context3["catch"](15);
            QRUrl = "https://cdn.wanwudezhi.com/static/web-static/image/84883f67019e8d7d29b4080e743957df.png";
            return _context3.abrupt("return", QRUrl);

          case 26:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3, null, [[15, 22]]);
  }));
  return _getH5Code.apply(this, arguments);
}

function getShareCodeImg() {
  return _getShareCodeImg.apply(this, arguments);
}

function _getShareCodeImg() {
  _getShareCodeImg = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4() {
    var options,
        shareOptions,
        skipLogin,
        _options$pushTarget,
        pushTarget,
        _options$style,
        style,
        _options$codeType,
        codeType,
        _args4 = arguments;

    return regeneratorRuntime.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            options = _args4.length > 0 && _args4[0] !== undefined ? _args4[0] : {};
            // options.shareOptions = Share.getCacheShareOptions(options.shareOptions);
            shareOptions = options.shareOptions, skipLogin = options.skipLogin, _options$pushTarget = options.pushTarget, pushTarget = _options$pushTarget === void 0 ? 'xcx' : _options$pushTarget, _options$style = options.style, style = _options$style === void 0 ? {} : _options$style, _options$codeType = options.codeType, codeType = _options$codeType === void 0 ? 0 : _options$codeType;

            if (!(codeType === 2)) {
              _context4.next = 6;
              break;
            }

            return _context4.abrupt("return", getQrCodeByStyle(options));

          case 6:
            if (!(codeType === 0)) {
              _context4.next = 10;
              break;
            }

            return _context4.abrupt("return", getH5Code(shareOptions, style));

          case 10:
            return _context4.abrupt("return", getXcxCode(shareOptions, skipLogin, pushTarget));

          case 11:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _getShareCodeImg.apply(this, arguments);
}

export default getH5Code;
export { getH5Code, getQrCodeByStyle, getShareCodeImg, getXcxCode };
