// @wanwu/base-fn/lib/image/ImageUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkIsCdnUrl = exports.getImageWithWatermark = exports.getOriginImage = exports.getRectangleImageWithHeight = exports.getRectangleImageWithWidth = exports.getRectangleImage = exports.getSquareImage = exports.getSmallImage = exports.checkSupportWebP = void 0;
/**
 * Create by changsheng on 2019-04-03 11:37
 */
var index_1 = require("../fn/index");
var isSupportWebP;
var cdnUrlPattern = /^http(s?):\/\/cdn\.wanwudezhi\.com/;
/**
 * 判断是否支持webp
 * @return {boolean}
 */
function checkSupportWebP() {
    if (isSupportWebP) {
        return isSupportWebP;
    }
    try {
        isSupportWebP = (document.createElement('canvas').toDataURL('image/webp').indexOf('data:image/webp') == 0);
        return isSupportWebP;
    }
    catch (err) {
        return false;
    }
}
exports.checkSupportWebP = checkSupportWebP;
// export const isSupportWebP = checkSupportWebP();
/**
 * 取400*400的方图
 * @param url
 * @param {string} format
 * @return {string}
 */
function getSmallImage(url, format) {
    return getSquareImage(url, 400, format);
}
exports.getSmallImage = getSmallImage;
/**
 * 取变成一样的图
 * @param url 图片地址
 * @param length 边长
 * @param {string} format 默认转成jpg
 * @return {string}
 */
function getSquareImage(url, length, format) {
    return getRectangleImage(url, { width: length, height: length, cut: true, format: format });
}
exports.getSquareImage = getSquareImage;
/**
 * 取矩形图，需要传长宽
 * @param url
 * @param {{cut?: boolean; width?: number; height?: number; format?: string}} options
 * @return {string}
 */
function getRectangleImage(url, options) {
    if (!url) {
        return url;
    }
    if (!cdnUrlPattern.test(url)) {
        return url;
    }
    if (!options) {
        options = {};
    }
    var originUrl = getOriginImage(url);
    var mode = '2';
    if (options.cut) {
        mode = '1';
    }
    // 获取文件默认后缀
    var suffix = originUrl.replace(/^.*\.(\w+)$/, '$1');
    // 导出最终的图片链接
    var videoReg = /(mp4|rm(vb)?|flv|mp(e)?(g)?|(n)?avi|wmv|m3u8|mov|mtv|asf|dat|mov|3gp|amv|dmv|m2v|vob|f4v|vcd)/;
    var offsetReg = /^.*\/offset\/(\w+)/;
    // format
    var format;
    // 处理ios新版手机的新的图片格式的问题
    if (suffix.toUpperCase() === 'HEIC' || suffix.toUpperCase() === 'HEIF') {
        format = 'jpg';
    }
    else {
        format = checkSupportWebP() ? 'webp' : (options.format || (videoReg.test(suffix.toLowerCase()) ? 'png' : suffix) || 'jpg');
    }
    // imageView2Src后缀
    var imageView2Src;
    if (options.width && options.height) {
        imageView2Src = "imageView2/" + mode + "/w/" + options.width + "/h/" + options.height + "/format/" + format;
    }
    else if (options.width) {
        imageView2Src = "imageView2/" + mode + "/w/" + options.width + "/format/" + format;
    }
    else if (options.height) {
        imageView2Src = "imageView2/" + mode + "/h/" + options.height + "/format/" + format;
    }
    else {
        imageView2Src = "imageView2/" + mode + "/format/" + format;
    }
    if (videoReg.test(suffix.toLowerCase())) {
        // 视频图片
        var _offset = void 0;
        if (index_1.typeYalidator.isEmpty(options.offset)) {
            if (offsetReg.test(url)) {
                _offset = url.replace(/^.*\/offset\/(\w+)/, '$1');
            }
            else {
                _offset = 0;
            }
        }
        else {
            _offset = options.offset;
        }
        return originUrl + "?vframe/png/offset/" + _offset + "|" + imageView2Src;
    }
    else {
        // 普通cdn图片
        return originUrl + "?" + imageView2Src;
    }
}
exports.getRectangleImage = getRectangleImage;
/**
 * 取矩形图，制定宽度，不裁剪
 * @param url 图片url
 * @param width 宽度
 * @param {string} format 后缀
 * @return {string}
 */
function getRectangleImageWithWidth(url, width, format) {
    return getRectangleImage(url, { width: width, format: format });
}
exports.getRectangleImageWithWidth = getRectangleImageWithWidth;
/**
 * 取矩形图，制定宽度，不裁剪
 * @param url 图片url
 * @param height 高度
 * @param {string} format 后缀
 * @return {string}
 */
function getRectangleImageWithHeight(url, height, format) {
    return getRectangleImage(url, { height: height, format: format });
}
exports.getRectangleImageWithHeight = getRectangleImageWithHeight;
function getOriginImage(url, watermark) {
    if (watermark === void 0) { watermark = false; }
    if (!checkIsCdnUrl(url)) {
        return url;
    }
    return url && url.split('?')[0] || '';
}
exports.getOriginImage = getOriginImage;
function getImageWithWatermark(url, watermarkString) {
    if (!checkIsCdnUrl(url)) {
        return url;
    }
    var originUrl = getOriginImage(url);
    if (originUrl) {
        var watermark = watermarkString || 'watermark/1/image/aHR0cDovL3BvcnRhbC1vcmlnaW4ud2Fud3VkZXpoaS5jb20vd2F0ZXJtYXJrX3YxLnBuZw==/dissolve/100/gravity/SouthEast/dx/10/dy/10';
        return originUrl + "?" + watermark;
    }
    return originUrl;
}
exports.getImageWithWatermark = getImageWithWatermark;
function checkIsCdnUrl(url) {
    return cdnUrlPattern.test(url);
}
exports.checkIsCdnUrl = checkIsCdnUrl;
