// @wanwu/im-encryption/src/constant/security-sign-constants.js

/**
 * sign
 */
const SignParamEnums = {
    // key相关
    SIGNSALT: 'e680d60e7e6bd5931cb46d30c91d6d0d', // key的混淆参数
    SIGNSECRETKEY: 'bbcc71f7b26a82ea97196366558a8ef0', // 签名密钥
    // headerkey名称
    SIGNEDHEADERSPARAMS: 'Signed-Headers',
    SIGNEDHEADERSAUTHORIZATION: 'Authorization',

    // 签名算法
    ALGORITHM: 'WWDZ-HMAC-SHA256',

    // 必须参与签名参数
    SIGNEDPARAM_METHOD: 'method',
    SIGNEDPARAM_APIPATH: 'apiPath',
    SIGNEDPARAM_SID: 'sId',
    SIGNEDPARAM_TIMESTAMP: 'timestamp',
    SIGNEDPARAM_APPVERSION: 'appVersion',
    SIGNEDPARAM_PAYLOAD: 'payload',
    SIGNEDPARAM_SIGNVERSION: 'signVersion',

    SIGNEDPARAM_DEVICETOKEN: 'deviceToken',


    SIGNEDPARAM_END: 'wwdz_request',

    // authorizationHeader参数名称
    AUTHORIZATIONHEADER_SIGNEDHEADERS: 'SignedHeaders',
    AUTHORIZATIONHEADER_ALGORITHM: 'Algorithm',
    AUTHORIZATIONHEADER_SIGNATURE: 'Signature',
    AUTHORIZATIONHEADER_CREDENTIAL: 'Credential',

};

const SignCodeEnums = {

    // 成功返回参数
    PASS: '0000',

    // 生成sign和校验sign公共返回错误码 0XXX
    SIGN_PARAMS_ERROR: '0001',

    // 生成sign返回错误码 1XXX

    properties: {
        '0000': {
            message: '签名通过',
        },
        '0001': {
            message: 'SignParams参数未传或者为空',
        },
    },

};
module.exports = {
    SignCodeEnums,
    SignParamEnums,
};
