// @wanwu/wx-sdk/src/classWxSdk.js
/*
 * @Author: 锦阳
 * @Create: 2021年02月05日
 */
/**
 * Create by changsheng on 2019-04-19 00:40
 */

import { ua, fn } from '@wanwu/base-fn';
import request from '@wanwu/base-request';
import RC from '@wanwu/resource-center';
import inject from '@wanwu/inject-script';
/**
 * 微信公众号配置信息
 * @typedef {Object} wxAppMap
 * @property {string} domain 域名
 * @property {string} appId 公众号appId
 * @property {string} userAppId 用户登录用appId
 * @property {number} appType 公众号类型
 */

/**
  * 微信配置参数
  * @typedef wxConfigParams
  * @property {string} nonceStr
  * @property {} signature
  * @property {number} timestamp
  */
const jweixinUrl = 'https://cdn.wanwudezhi.com/static/web-static/text/775529c69d2d5632895cc05e924780bb/jweixin-1.6.0.js';
/** wxSdk */
class WxSdk {
    /** 微信JS Sdk的appId */
    wxAppId= 'wxb2de1255783f7c70';
    /** 微信JS APi列表 */
    jsApiList= [
        'downloadImage',
        'chooseWXPay',
        'openAddress',
        'updateAppMessageShareData',
        'updateTimelineShareData',
        'onMenuShareAppMessage',
        'onMenuShareTimeline',
        'hideAllNonBaseMenuItem',
        'showAllNonBaseMenuItem',
        'hideMenuItems',
        'showMenuItems',
        'previewImage',
        'getNetworkType',
        'closeWindow',
        'scanQRCode',
        'openLocation',
        'chooseImage',
        'startRecord',
        'stopRecord',
        'translateVoice',
    ];

    /* 微信开放标签列表 */
    openTagList= [
        'wx-open-launch-app',
    ];

    _isWxReady=null;
    _isMiniProgram= false;
    /** 微信配置接口 */
    configApi= 'https://doc.wanwudezhi.com/api/v1/wx/event/ticketNew';
    /**
     * 初始化微信sdk配置
     * @returns {Promise<void>}
     */
    initializeConfig() {
        this.wxConfigBatch = fn.debounce(this.wxConfig.bind(this), 500, false);

        // iOS中只有验证失败的时候，需要重新验证
        if (ua.isIOS() && this._isWxReady !== null) {
            this._isWxReady.then(() => {
                window.wx.error(() => {
                    this.wxConfigBatch();
                });
            });
            return this._isWxReady;
        }

        // Android每次都进行重新验证
        if (ua.isWechat()) {
            this._isWxReady = this.wxConfig();
        }

        // 注入所有的微信Api方法
        this.jsApiList.forEach(key => {
            if (ua.isWechat()) {
                WxSdk[key] = (options) => {
                    this.wxReady().then(() => {
                        window.wx[key] && window.wx[key](options);
                    });
                };
            } else {
                WxSdk[key] = (options) => {
                    options.fail && options.fail({
                        errMsg: '请在微信中打开',
                    });
                };
            }
        });
        return this._isWxReady;
    }

    /**
     * 获取微信公众号相关的配置
     * @returns {Promise<wxAppMap>}
     */
    async getWxAppMap() {
        try {
            const appIdMapList = await RC.getList(157);
            const domain = location.hostname;
            return appIdMapList.find((appIdMap) => {
                return appIdMap.domain === domain;
            }) || {};
        } catch (e) {
            return {};
        }
    }

    /**
     * 获取微信公众号AppId
     * @returns {Promise<string>}
     */
    async getWxAppId() {
        const appIdMap = await this.getWxAppMap();
        return appIdMap.appId || this.wxAppId;
    }

    /**
     * 获取微信公众号类型
     * @returns {Promise<number>}
     */
    async getWxAppType() {
        const appIdMap = await this.getWxAppMap();
        return appIdMap.appType || 4;
    }

    /**
     * 注入微信SDK
     * @returns {Promise<void>}
     */
    injectWeiXinSDK() {
        if (!window.wx) {
            return inject(jweixinUrl);
        }
    }

    /**
     * 获取微信配置
     * @returns {Promise<wxConfigParams>}
     */
    async getWxConfigParams() {
        const url = window.location.href.split('#')[0];
        const wxAppId = await this.getWxAppId();
        return await request.post(this.configApi, { url, app: wxAppId }, { withoutToken: true });
    }

    /**
     * 初始化微信配置
     */
    async wxConfig() {
        const [wxAppId, wxConfigParams] = await Promise.all([
            this.getWxAppId(),
            this.getWxConfigParams(),
            this.injectWeiXinSDK(),
        ]);
        const { nonceStr, signature, timestamp } = wxConfigParams;

        window.wx.config({
            debug: false, // 开启调试模式,调用的所有api的返回值会在客户端alert出来，若要查看传入的参数，可以在pc端打开，参数信息会通过log打出，仅在pc端时才会打印。
            appId: wxAppId, // 必填，公众号的唯一标识
            timestamp, // 必填，生成签名的时间戳
            nonceStr, // 必填，生成签名的随机串
            signature, // 必填，签名
            jsApiList: this.jsApiList, // 必填，需要使用的JS接口列表
            openTagList: this.openTagList,
        });
    }

    /**
     * 微信sdk初始化
     * @returns {Promise<void>}
     */
    async wxReady() {
        return new Promise((resolve, reject) => {
            if (ua.isWechat()) {
                this._isWxReady.then(() => {
                    window.wx.ready(() => {
                        resolve();
                    });

                    window.wx.error(() => {
                        if (ua.isIOS()) {
                            this.wxConfigBatch();
                        }
                    });
                });
            } else {
                reject(new Error('请在微信中打开'));
            }
        });
    }

    /**
     * 是否是小程序环境
     * @returns {Promise<boolean>}
     */
    isMiniProgram() {
        return new Promise((resolve, reject) => {
            this.wxReady().then(() => {
                if (window.wx.miniProgram && window.wx.miniProgram.getEnv) {
                    window.wx.miniProgram.getEnv((res) => {
                        this._isMiniProgram = res.miniprogram;
                        resolve(res.miniprogram);
                    });
                } else {
                    this._isMiniProgram = false;
                    resolve(false);
                }
            }).catch(() => {
                this._isMiniProgram = false;
                resolve(false);
            });
        });
    }

    /**
     * 跳转微信小程序页面
     * @param {string} url 小程序路径
     */
    wxMiniProgramNavigateTo(url) {
        return new Promise((resolve, reject) => {
            this.wxReady().then(() => {
                window.wx.miniProgram.navigateTo({
                    url,
                    success(res) {
                        resolve(res);
                    },
                    fail(res) {
                        reject(res);
                    },
                });
            });
        });
    }

    /**
    * 修改微信appId
    * @param {string} wxAppId 微信appId
    */
    setWxAppId(wxAppId) {
        this.wxAppId = wxAppId;
    }

    /**
     * 修改配置接口
     * @param {string} configApi 配置接口
     */
    setConfigApi(configApi) {
        this.configApi = configApi;
    }
}
// 初始化交给使用的人
// WxSdk.isMiniProgram();
// WxSdk.initializeConfig();

// const instance = new WxSdk();
export default new WxSdk();
