// @wanwu/wx-sdk/src/index.js
/**
 * Create by changsheng on 2019-04-19 00:40
 */

import { ua, fn } from '@wanwu/base-fn';
import request from '@wanwu/base-request';
import RC from '@wanwu/resource-center';
import inject from '@wanwu/inject-script';

const jweixinUrl = 'https://cdn.wanwudezhi.com/static/web-static/text/775529c69d2d5632895cc05e924780bb/jweixin-1.6.0.js';
const WxSdk = {
    wxAppId: 'wxb2de1255783f7c70',
    jsApiList: [
        'downloadImage',
        'chooseWXPay',
        'openAddress',
        'updateAppMessageShareData',
        'updateTimelineShareData',
        'onMenuShareAppMessage',
        'onMenuShareTimeline',
        'hideAllNonBaseMenuItem',
        'showAllNonBaseMenuItem',
        'hideMenuItems',
        'showMenuItems',
        'previewImage',
        'getNetworkType',
        'closeWindow',
        'scanQRCode',
        'openLocation',
        'chooseImage',
        'wx-open-launch-weapp',
        'startRecord',
        'stopRecord',
        'translateVoice',
    ],
    openTagList: [
        'wx-open-launch-app',
        'wx-open-launch-weapp',
    ],
    _isWxReady: null,
    _isMiniProgram: false,
    configApi: 'https://doc.wanwudezhi.com/api/v1/wx/event/ticketNew',

    initializeConfig() {
        this.wxConfigBatch = fn.debounce(this.wxConfig.bind(this), 500, false);

        // iOS中只有验证失败的时候，需要重新验证
        if (ua.isIOS() && this._isWxReady !== null) {
            this._isWxReady.then(() => {
                window.wx.error(() => {
                    this.wxConfigBatch();
                });
            });
            return this._isWxReady;
        }

        // Android每次都进行重新验证
        if (ua.isWechat()) {
            this._isWxReady = this.wxConfig();
        }

        // 注入所有的微信Api方法
        this.jsApiList.forEach(key => {
            if (ua.isWechat()) {
                WxSdk[key] = (options) => {
                    this.wxReady().then(() => {
                        window.wx[key] && window.wx[key](options);
                    });
                };
            } else {
                WxSdk[key] = (options) => {
                    options.fail && options.fail({
                        errMsg: '请在微信中打开',
                    });
                };
            }
        });
        return this._isWxReady;
    },

    async getWxAppMap() {
        try {
            const appIdMapList = await RC.getList(157);
            const domain = location.hostname;
            return appIdMapList.find((appIdMap) => {
                return appIdMap.domain === domain;
            }) || {};
        } catch (e) {
            return {};
        }
    },

    async getWxAppId() {
        const appIdMap = await this.getWxAppMap();
        return appIdMap.appId || this.wxAppId;
    },

    async getWxAppType() {
        const appIdMap = await this.getWxAppMap();
        return appIdMap.appType || 4;
    },

    injectWeiXinSDK() {
        if (!window.wx) {
            return inject(jweixinUrl);
        }
    },

    async getWxConfigParams() {
        const url = window.location.href.split('#')[0];
        const wxAppId = await this.getWxAppId();
        return await request.post(this.configApi, { url, app: wxAppId }, { withoutToken: true });
    },

    async wxConfig() {
        const [wxAppId, wxConfigParams] = await Promise.all([
            this.getWxAppId(),
            this.getWxConfigParams(),
            this.injectWeiXinSDK(),
        ]);
        const { nonceStr, signature, timestamp } = wxConfigParams;

        window.wx.config({
            debug: false, // 开启调试模式,调用的所有api的返回值会在客户端alert出来，若要查看传入的参数，可以在pc端打开，参数信息会通过log打出，仅在pc端时才会打印。
            appId: wxAppId, // 必填，公众号的唯一标识
            timestamp, // 必填，生成签名的时间戳
            nonceStr, // 必填，生成签名的随机串
            signature, // 必填，签名
            jsApiList: this.jsApiList, // 必填，需要使用的JS接口列表
            openTagList: this.openTagList,
        });
    },

    async wxReady() {
        return new Promise((resolve, reject) => {
            if (ua.isWechat()) {
                this._isWxReady.then(() => {
                    window.wx.ready(() => {
                        resolve();
                    });

                    window.wx.error(() => {
                        if (ua.isIOS()) {
                            this.wxConfigBatch();
                        }
                    });
                });
            } else {
                reject(new Error('请在微信中打开'));
            }
        });
    },

    isMiniProgram() {
        return new Promise((resolve, reject) => {
            this.wxReady().then(() => {
                if (window.wx.miniProgram && window.wx.miniProgram.getEnv) {
                    window.wx.miniProgram.getEnv((res) => {
                        this._isMiniProgram = res.miniprogram;
                        resolve(res.miniprogram);
                    });
                } else {
                    this._isMiniProgram = false;
                    resolve(false);
                }
            }).catch(() => {
                this._isMiniProgram = false;
                resolve(false);
            });
        });
    },

    wxMiniProgramNavigateTo(url) {
        return new Promise((resolve, reject) => {
            this.wxReady().then(() => {
                window.wx.miniProgram.navigateTo({
                    url,
                    success(res) {
                        resolve(res);
                    },
                    fail(res) {
                        reject(res);
                    },
                });
            });
        });
    },

    wxMiniProgramRedirectTo(url) {
        return new Promise((resolve, reject) => {
            this.wxReady().then(() => {
                window.wx.miniProgram.redirectTo({
                    url,
                    success(res) {
                        resolve(res);
                    },
                    fail(res) {
                        reject(res);
                    },
                });
            });
        });
    },

    setWxAppId(wxAppId) {
        this.wxAppId = wxAppId;
    },
    setConfigApi(configApi) {
        this.configApi = configApi;
    },
};
// 初始化交给使用的人
// WxSdk.isMiniProgram();
// WxSdk.initializeConfig();

// const instance = new WxSdk();

export default WxSdk;
