// @wanwu/base-vue-location-link/lib/index.es.js
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.regexp.exec';
import 'core-js/modules/es.string.replace';
import wLocation from '@wanwu/w-location';
export { throughParams, default as wLocation } from '@wanwu/w-location';
import { normalizeComponent } from '@wanwu/vue-runtime-helpers';

var script = {
  name: 'LocationLink',
  props: {
    to: {
      type: [String, Object]
    },
    replace: {
      type: Boolean
    }
  },
  methods: {
    locationTo: function locationTo() {
      if (!this.$location) {
        this.$location = wLocation;
      }

      this.$location[this.replace ? 'replace' : 'push'](this.to);
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('span', {
    staticClass: "component__locationLink locationLink",
    on: {
      "click": _vm.locationTo
    }
  }, [_vm._t("default")], 2);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = undefined;
/* scoped */

var __vue_scope_id__ = undefined;
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject */

/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, undefined, undefined, undefined);

__vue_component__.install = function (Vue, options) {
  Vue.use(wLocation, options);
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
