// @wanwu/base-vue-location-link/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('core-js/modules/es.function.name');
require('core-js/modules/es.regexp.exec');
require('core-js/modules/es.string.replace');
var wLocation = require('@wanwu/w-location');
var wLocation__default = _interopDefault(wLocation);
var vueRuntimeHelpers = require('@wanwu/vue-runtime-helpers');

var script = {
  name: 'LocationLink',
  props: {
    to: {
      type: [String, Object]
    },
    replace: {
      type: Boolean
    }
  },
  methods: {
    locationTo: function locationTo() {
      if (!this.$location) {
        this.$location = wLocation__default;
      }

      this.$location[this.replace ? 'replace' : 'push'](this.to);
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('span', {
    staticClass: "component__locationLink locationLink",
    on: {
      "click": _vm.locationTo
    }
  }, [_vm._t("default")], 2);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = undefined;
/* scoped */

var __vue_scope_id__ = undefined;
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject */

/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = vueRuntimeHelpers.normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, undefined, undefined, undefined);

__vue_component__.install = function (Vue, options) {
  Vue.use(wLocation__default, options);
  Vue.component(__vue_component__.name, __vue_component__);
};

Object.defineProperty(exports, 'throughParams', {
    enumerable: true,
    get: function () {
        return wLocation.throughParams;
    }
});
exports.wLocation = wLocation__default;
exports.default = __vue_component__;
