// @wanwu/vue-nine-h5-SingleBanner/config.js
export default [
    {
        type: 'ImageLinkItem',
        title: '链接+图片',
        name: 'linkImg',
        props: {
            url: {
                name: 'url',
                placeholder: '请上传/输入图片',
                maxSize: 4,
                type: 'png/jpg',
                canEdit: true,
                hasPreview: false,
            },
            link: {
                name: 'link',
                placeholder: '请输入跳转链接',
                title: '请输入跳转链接',
            },
        },
        rules: [
            {
                required: true,
                trigger: [
                    'blur',
                    'change',
                ],
            },
            {
                validator: (rule, value, callback) => {
                    if (!value.url) {
                        callback(new Error('图片url不能为空'));
                    }
                },
            },
        ],
    },
    {
        type: 'Slider',
        title: '高度',
        name: 'height',
        default: null,
        props: {
            min: 0,
            max: 500,
        },
    },
    {
        type: 'Slider',
        title: '上边距',
        name: 'marginTop',
        default: null,
        props: {
            min: 0,
            max: 200,
        },
    },
    {
        type: 'Slider',
        title: '下边距',
        name: 'marginBottom',
        default: null,
        props: {
            min: 0,
            max: 200,
        },
    },
];
