// @wanwu/base-logger/src/util/types.ts
export interface LoggerInfoTypes {
    userAgent?: any;
    basic?: any;
    ext?: object;
    eventId?: number;
    rtpCnt?: string;
    event?: string;
    url?: string;
    refer?: string;
    rtpRefer?: any;
    traceLink?: {
        [key: string]: any;
    }
}

export interface NewLoggerInfoTypes {
    url?: string;
    rtpRefer?: string;
    rtpCnt?: string;
    browser?: any;
    basic?: any;
    event?: string;
    ext?: object;
    refer?: string;
    traceLink?: {
        [key: string]: any;
    }
}

export interface ConfigTypes {
    LogUrl: string;
    NewLogUrl: string;
    CORS__LogUrl: string;
    NEW__CORS__LogUrl: string;
    UserInfoKey: string;
    rtp_cnt_a: string;
    rtp_cnt_b: string;
    eventId: number;
    platform: string;
    ignorePageRegExp: RegExp[];
    send_interval: number;
    one_send_max_length: number;
    filters: any[];
    env_production_host: RegExp;
    traceLink: {
        [key: string]: any;
    }
}

export declare interface Window {
    $loggerData?: any
}
