// @wanwu/base-mall-logger/lib/index.es.js
import 'core-js/modules/es.symbol';
import 'core-js/modules/es.symbol.description';
import 'core-js/modules/es.symbol.iterator';
import 'core-js/modules/es.array.concat';
import 'core-js/modules/es.array.for-each';
import 'core-js/modules/es.array.from';
import 'core-js/modules/es.array.is-array';
import 'core-js/modules/es.array.iterator';
import 'core-js/modules/es.array.join';
import 'core-js/modules/es.array.map';
import 'core-js/modules/es.array.slice';
import 'core-js/modules/es.date.now';
import 'core-js/modules/es.date.to-string';
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.object.assign';
import 'core-js/modules/es.object.keys';
import 'core-js/modules/es.object.to-string';
import 'core-js/modules/es.regexp.exec';
import 'core-js/modules/es.regexp.to-string';
import 'core-js/modules/es.string.iterator';
import 'core-js/modules/es.string.search';
import 'core-js/modules/web.dom-collections.for-each';
import 'core-js/modules/web.dom-collections.iterator';
import 'core-js/modules/web.timers';
import { fn as fn$1, ua as ua$1 } from '@wanwu/base-fn';
import { logE } from '@wanwu/base-logger';
import uaParser from '@wanwu/ua-parse';
import mySentry from '@wanwu/sentry';
import axios from '@wanwu/base-request';
import User from '@wanwu/base-user';

function _createForOfIteratorHelper(o) { if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (o = _unsupportedIterableToArray(o))) { var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var it, normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
var fn = fn$1;
var ua = ua$1;
function initLogger() {
  var userInfo = User.getUserInfo() || {};
  userInfo.id = userInfo.userId || '-1';
  userInfo.username = userInfo.unick || '';
  var systemInfo = uaParser(ua.getUserAgent());
  var systemInfoTags = {};
  Object.keys(systemInfo).forEach(function (key) {
    var item = systemInfo[key];
    systemInfoTags[key + '-uaParser'] = typeof item === 'string' ? item : Object.keys(item).map(function (keyStr) {
      var itemStr = item[keyStr] || '';
      return itemStr;
    }).join(' ');
  });
  mySentry.setUserOrTag({
    user: userInfo,
    tags: systemInfoTags,
    content: {
      key: 'systemInfo',
      value: systemInfo
    }
  });
}
var Logger = {
  msgQueue: [],
  timer: null,
  logP: function logP() {
    if (fn.getQueryString('channel')) {
      axios.get("topic/count?channel=".concat(encodeURIComponent(fn.getQueryString('channel'))))["catch"](function (err) {
        mySentry.logError(err);
      });
    }
  },
  logE: function logE$1(eventId, extra) {
    logE(eventId, extra);
  },
  sendMsg: function sendMsg(error) {
    var _this = this;

    var params = {
      msg: error.message,
      stack: error.stack,
      userInfo: User.getUserInfo(),
      ua: navigator.userAgent,
      timestamp: Date.now(),
      url: location.href,
      referer: document.referrer
    };
    this.msgQueue.push(params);
    clearTimeout(this.timer);
    this.timer = setTimeout(function () {
      _this.requestLog();
    }, 0);
  },
  requestLog: function requestLog() {
    var _this2 = this;

    var tmpData = this.msgQueue;
    this.msgQueue = []; // axios({
    //     url: 'https://api.wanwudezhi.com/media/log/add',
    //     data: tmpData.map(item => {
    //         return JSON.stringify(item);
    //     }),
    //     method: 'POST',
    //     headers: {
    //         [Config.AuthorizationKey]: '',
    //     },
    // })
    //     .then(() => {
    //     })
    //     .catch((err) => {
    //         mySentry.logError(err);
    //         this.msgQueue = this.msgQueue.concat(tmpData);
    //     });

    axios.post('https://api.wanwudezhi.com/media/log/add', tmpData.map(function (item) {
      return JSON.stringify(item);
    }), {
      withoutToken: true
    }).then(function () {})["catch"](function (err) {
      mySentry.logError(err);
      _this2.msgQueue = _this2.msgQueue.concat(tmpData);
    });
  }
};
function VueErrorHandler(err, vm, info) {
  Logger.sendMsg(err);
}
function VueWarnHandler(msg, vm, info) {
  Logger.sendMsg({
    message: msg,
    stack: info
  });
}
var throughParams = ['__env', 'inviteCode', 'token', '__debug', '__useip', '__useport', 'shareUserId', 'shareTime', '__source_channel', '__miniprogramType'];
function getRouteLocation(url, params) {
  var urlSearch = location.search.slice(1);
  var urlParams = fn.parse(urlSearch);
  var pickParams = fn.pick(urlParams, throughParams); // stringifyUrl: 将对象作为search参数，拼装成url

  return fn.stringifyUrl(url, Object.assign({}, pickParams, params));
}
function getRouteQuery(to, from) {
  var fromQuery = from.query;
  var toQuery = to.query; // 进行从原对象中过滤需要的属性，返回过滤的对象

  var pickedParams = fn.pick(fromQuery, throughParams);

  if (toQuery.inviteCode && Array.isArray(toQuery.inviteCode)) {
    var _iterator = _createForOfIteratorHelper(toQuery.inviteCode),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var item = _step.value;

        if (item) {
          toQuery.inviteCode = item;
          break;
        }
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
  }

  return Object.assign({}, toQuery, pickedParams);
}

export default Logger;
export { VueErrorHandler, VueWarnHandler, getRouteLocation, getRouteQuery, initLogger };
