// @wanwu/nine-dll/lib/index.js
/**
 *file: conmmonEntry.js
 *@Author: wxm
 *@Date: 2020-04-03 15:36
 **/
import Vue from 'vue';
import flexible from '@wanwu/flexible';
import '@wanwu/base-font-wwdz'; // import VConsole from 'vconsole';

import * as wanwu from '@wanwu/base-fn';
import locationPluginTimestamp from '@wanwu/bus-sdk-location-plugin-timestamp';
import Location from '@wanwu/w-location';
import lazyload from '@wanwu/base-vue-directive-lazy';
import locationReplace, { PREVENTCODE } from '@wanwu/location-plugin-replace';
import lazyComponent from '@wanwu/base-vue-lazy-component/lib/index';
import LocationLink, { wLocation } from '@wanwu/base-vue-location-link';
import { logE } from '@wanwu/base-logger';
import Toast from '@wanwu/base-vue-toast';
import '@wanwu/fastclick';
import '@wanwu/base-fn/lib/polyfill/Date';
import Logger, { initLogger } from '@wanwu/base-mall-logger';
import logger from './plugins/logger';
import request from './plugins/request';
import kl_request from './plugins/kl_request';
import User from './User';
import ui from './plugins/ui';
import Wrapper from './Wrapper';
import VariablesProvider from './VariablesProvider';
import mySentry from '@wanwu/sentry';
import logDirective from '@wanwu/base-vue-directive-log';

if (process.env.NODE_ENV === 'production') {
  console.log('------sentry init--------');
  var feEnv = /^pre-/.test(window.location.hostname) ? 'pre' : /^dev-/.test(window.location.hostname) ? 'dev' : 'prod';
  var apiEnv = wanwu.fn.getQueryString('__env') || 'prod';
  mySentry.init({
    dsn: 'https://1b75212d0a814937915a5fb1a6f47ecd@sentry.wanwudezhi.com/8',
    environment: "feEnv-".concat(feEnv, "-apiEnv-").concat(apiEnv, "-Sentry-SDK"),
    vue: Vue,
    window: window,
    env: 'vue',
    callback: function callback() {
      initLogger();
    }
  });
}

window.RTP_PARAMS = {
  rtp_cnt_a: 'bw6',
  // 商城
  rtp_cnt_b: 'w0',
  // h5
  env_production_host: /^(pre-|m\.(wanwudezhi|wwdz3|wwdz2)\.com)/
};

window.onerror = function (message, source, l, c, error) {
  console.log(error);
  Logger.sendMsg(error || {});
};

export { Wrapper, VariablesProvider };
export default (function (env) {
  // [MOD]移动到main.js中运行
  // if (env !== 'preview') {
  //     // 记录页面PV
  //     logP();
  // }
  var locationOptions = {
    errorCallback: function errorCallback(err) {
      var path = wanwu.fn.getObjDeepValue(err, 'to.path');
      var query = wanwu.fn.getObjDeepValue(err, 'to.query');
      var appPath = wanwu.fn.getObjDeepValue(err, 'config.appUrl');

      switch (err.code) {
        case PREVENTCODE.appVersionPrevent:
          Vue.toast('当前版本过低，请升级App后重试！');
          break;

        case PREVENTCODE.appPrevent:
          switch (path) {
            case '/good/upload':
              Vue.toast('请前往【拍卖管理】发布拍卖活动商品');
              break;
          }

          break;

        case PREVENTCODE.xcxPrevent:
          break;

        case PREVENTCODE.h5Prevent:
          switch (path) {
            case '/live/room':
              Vue.dialog.confirm({
                message: '暂不支持查看，打开APP查看更流畅！',
                confirmButtonText: '打开APP'
              }).then(function () {
                Vue.callapp.open({
                  path: query.liveingFlag + '' === '1' ? 'live_room' : 'live_preview',
                  param: query
                });
              })["catch"](function () {});
              break;

            case '/im/support':
              Vue.dialog.confirm({
                message: '暂不支持查看，打开APP查看更流畅！',
                confirmButtonText: '打开APP'
              }).then(function () {
                Vue.callapp.open({
                  path: "customer7?param=".concat(JSON.stringify(query)),
                  param: query
                });
              })["catch"](function () {});
              break;

            default:
              Vue.dialog.confirm({
                message: '暂不支持查看，打开APP查看更流畅！',
                confirmButtonText: '打开APP'
              }).then(function () {
                Vue.callapp.open({
                  path: (appPath || path).replace(/^\//, ''),
                  param: query
                });
              })["catch"](function () {});
          }

          break;
      }
    }
  };
  wLocation.use(locationReplace); // 添加

  wLocation.use(function (to) {
    var isNoOrigin = !to.origin || to.origin === window.location.origin;

    if (isNoOrigin && !/^\/(preview|page|redirect)\/./.test(to.path)) {
      to.origin = "http".concat(to.query.__env === 'pre' || to.query.__env === 'dev' || to.query.__env === 'develop' ? '' : 's', "://h5.wanwudezhi.com");

      if (!to.base) {
        to.base = '/mall-web';
      }
    }

    if (to.path === '/user/bindPhone' && (to.query.__env === 'pre' || to.query.__env === 'dev' || to.query.__env === 'develop')) {
      to.origin = 'http://h5.wanwudezhi.com';
    }

    return env === 'preview' && to.path !== '/user/bindPhone' ? null : to;
  });
  Vue.use(LocationLink, locationOptions);
  Vue.mixin({
    methods: wanwu.ImageUtils
  });
  Vue.use(lazyComponent, {
    limit: 200
  });
  Vue.use(Toast);
  Vue.use(lazyload, {
    loading: 'https://cdn.wanwudezhi.com/static/web-static/image/1cfb810cf69db4ea0e0bb3d7a64b71c6_555x555.png',
    error: 'https://cdn.wanwudezhi.com/static/web-static/image/fdaa64918a5a15e53fc1a87c9d870e61_555x555.png'
  });
  flexible(window, window['BaseLib'] || (window['BaseLib'] = {}));

  if (wanwu.fn.getQueryString('__debug') === wanwu.fn.getDebugCode()) {
    // const vConsole = import('vconsole')
    var loadVConsole = function loadVConsole() {
      return import('vconsole');
    };

    loadVConsole().then(function (module) {
      var vConsole = new module["default"](); // eslint-disable-line

      console.log(vConsole);
    })["catch"](function (err) {
      console.error('load vconsole error', err);
    });
  }

  Location.use(locationPluginTimestamp);
  window.wanwu = wanwu;
  Vue.use(request);
  Vue.use(kl_request);
  Vue.use(ui);
  Vue.use(logger); // TODO: 更新logE

  Vue.prototype.$logE = function (eventId, ext) {
    var parent = this;
    var moduleType, moduleName; // eslint-disable-next-line

    while (true) {
      parent = parent && parent.$parent;
      if (!parent) break;
      var config = parent.config;
      if (!config) continue;
      moduleType = config.componentName;
      moduleName = config.commonProps && config.commonProps.title;
      if (moduleType) break;
    }

    ext = ext || {};
    ext.moduleName = moduleName;
    ext.moduleType = moduleType;
    logE(eventId, ext);
  };

  Vue.use(logDirective);
  Vue.config.ignoredElements = ['wx-open-launch-weapp'];
  Vue.prototype.$location = Location;
  Vue.config.productionTip = false; // NOTE: 上报移动至main.js下mounted中

  if (['prerender', 'preview'].includes(env)) {
    User.updateToken()["catch"](console.log);
  }
});
