// @wanwu/base-vue-tabs-bar/lib/index.es.js
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.symbol';
import 'core-js/modules/es.array.filter';
import 'core-js/modules/es.array.for-each';
import 'core-js/modules/es.array.includes';
import 'core-js/modules/es.array.map';
import 'core-js/modules/es.array.splice';
import 'core-js/modules/es.number.constructor';
import 'core-js/modules/es.object.define-properties';
import 'core-js/modules/es.object.define-property';
import 'core-js/modules/es.object.get-own-property-descriptor';
import 'core-js/modules/es.object.get-own-property-descriptors';
import 'core-js/modules/es.object.keys';
import 'core-js/modules/es.string.includes';
import 'core-js/modules/web.dom-collections.for-each';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import 'regenerator-runtime/runtime';
import _asyncToGenerator from '@babel/runtime/helpers/esm/asyncToGenerator';
import { mapState } from 'vuex';
import ButtonX from '@wanwu/base-vue-bottom-x';
import Badge from '@wanwu/base-vue-badge';
import Occupying from '@wanwu/base-vue-occupying';
import { ua } from '@wanwu/base-fn';
import { bottomResource, activeGoodsType } from '@wanwu/resource-center';
import wLocation from '@wanwu/w-location';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
var script = {
  name: 'tabsBar',
  components: {
    ButtonX: ButtonX,
    // ShareBtn,
    Occupying: Occupying,
    Badge: Badge
  },
  props: {
    active: {
      type: Number,
      "default": 0
    },
    updateVideoNum: {
      type: Number,
      "default": 0
    }
  },
  data: function data() {
    return {
      list: [{
        index: 0,
        text: '首页',
        pagePath: '/home/index',
        activeIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/a35774df4a776fb786caaafff40bd976.png',
        normalIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/e91a08fb7c15969afdaf021b2b05024e.png'
      }, {
        index: 1,
        text: '分类',
        pagePath: '/category/index',
        activeIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/3ffdc055738941fb7659f29cdaf0e303.png',
        normalIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/896e039711b8c8df35aca1408c0843e1.png'
      }, {
        index: 3,
        text: '我的',
        pagePath: '/personal/center',
        activeIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/46c0159277c6eb8d4756bef6d988d847.png',
        normalIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/1f422779f30f6e35336598024b925cd0.png'
      }],
      activityList: null,
      isIPhoneX: false,
      // promo: Number(localStorage.getItem('promo')),
      localStorageKey: '_bottom_list' // associationSwitch: false, // 开关是否开启

    };
  },
  methods: {
    // onBtnClick(link, isCur) {
    //     if (link && !isCur) {
    //         const baseUrl = link === '/community/index' ? '/community-web' : '/mall-web';
    //         this.$location.push({
    //             base: baseUrl,
    //             path: link,
    //         });
    //     }
    // },
    onBtnClick: function onBtnClick(link, isCur) {
      if (link && !isCur) {
        var baseUrl = '';

        if (link === '/community/index') {
          baseUrl = '/community-web';
        } else if (window.location.pathname.includes('/community-web')) {
          baseUrl = '/mall-web';
        } else {
          baseUrl = '';
        }

        this.$location.push({
          base: baseUrl,
          path: link
        });
      }
    },
    // getActivityStatus() {
    //     this.$request.get('api/v1/mall/client/get').then(res => {
    //         // console.log(res);
    //         this.promo = res.promo;
    //         localStorage.setItem('promo', this.promo);
    //     }).catch(err => {
    //         console.error(err);
    //     });
    // },
    // GET /api/v1/mall/page/resource/bottom/xcx
    getBottomConfig: function getBottomConfig() {
      var _this = this;

      // bottomResource
      bottomResource().then(function (res) {
        if (res && res.detail && res.detail.length >= 3) {
          var list = res.detail.map(function (item, index) {
            return {
              index: index,
              text: item.name,
              pagePath: _this.list[index].pagePath,
              activeIcon: item.selectImg.url,
              normalIcon: item.img.url
            };
          }); // const list = this.list.map((item, index) => {
          //     const resItem = res.detail[index];
          //     return {
          //         index,
          //         text: resItem.name,
          //         pagePath: item.pagePath,
          //         activeIcon: resItem.selectImg.url,
          //         normalIcon: resItem.img.url,
          //     };
          // });

          localStorage.setItem(_this.localStorageKey, JSON.stringify(list));
          _this.activityList = list;
        } else {
          localStorage.removeItem(_this.localStorageKey);
          _this.activityList = null;
        }
      })["catch"](function (err) {
        _this.$logError(err);
      });
    },
    getShowCommunity: function getShowCommunity() {
      var _this2 = this;

      return _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var res, hideComminuty, communityConfig;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.prev = 0;
                _context.next = 3;
                return activeGoodsType();

              case 3:
                res = _context.sent;
                // this.associationSwitch = res.associationSwitch;
                // 只有res.associationSwitch值为false才隐藏，其他情况下都不隐藏
                hideComminuty = res.associationSwitch === false;
                communityConfig = {
                  index: 2,
                  text: '社区',
                  pagePath: '/community/index',
                  activeIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/1bd8ce1ef5e9e8242859d845f70c0376_90x90.png',
                  normalIcon: 'https://cdn.wanwudezhi.com/static/web-static/image/7e626cc188e8d8b119d20d7daceb8501_90x90.png'
                };
                if (!hideComminuty) _this2.list.splice(2, 0, communityConfig);
                localStorage.setItem(_this2.localStorageKey, JSON.stringify(_this2.list));
                _context.next = 13;
                break;

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](0);

                _this2.$logError(_context.t0);

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, null, [[0, 10]]);
      }))();
    }
  },
  computed: _objectSpread({}, mapState('community', ['socialCount'])),
  created: function created() {
    var _this3 = this;

    return _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
      var localConfig;
      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              if (!_this3.$location) _this3.$location = wLocation;
              _this3.isIPhoneX = ua.isIPhoneX(); // 这期暂时注释，先走本地数据
              // this.getBottomConfig();
              // 初次判断社区是否需要显示

              _context2.next = 4;
              return _this3.getShowCommunity();

            case 4:
              localConfig = localStorage.getItem(_this3.localStorageKey);

              if (localConfig) {
                _this3.activityList = JSON.parse(localConfig);
              } // this.promo = Number(localStorage.getItem('promo'));
              // this.getActivityStatus();


            case 6:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }))();
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "component__tabsBar tabsBar"
  }, [_c('ButtonX', {
    staticClass: "ButtonX"
  }, [_c('div', {
    staticClass: "item flex f-ai-c f-jc-sa"
  }, _vm._l(_vm.activityList || _vm.list, function (item, index) {
    return _c('button', {
      key: item.pagePath + index,
      staticClass: "item__btn btn flex f-fd-c f-ai-c",
      on: {
        "click": function click($event) {
          return _vm.onBtnClick(item.pagePath, item.index === _vm.active);
        }
      }
    }, [item.index === 2 && _vm.socialCount > 0 ? _c('Badge', {
      staticClass: "badge-box",
      attrs: {
        "isDot": ""
      }
    }) : _vm._e(), _vm._v(" "), _c('img', {
      directives: [{
        name: "lazy",
        rawName: "v-lazy",
        value: item.index === _vm.active ? item.activeIcon : item.normalIcon,
        expression: "item.index === active? item.activeIcon: item.normalIcon"
      }],
      staticClass: "item__btn-icon"
    }), _vm._v(" "), _c('div', {
      staticClass: "item__btn-text",
      "class": {
        active: item.index === _vm.active
      },
      style: {
        color: item.index === _vm.active ? item.selectFontColour : ''
      }
    }, [_vm._v(_vm._s(item.text))])], 1);
  }), 0)]), _vm._v(" "), _c('Occupying', {
    staticClass: "Occupying",
    attrs: {
      "height": _vm.isIPhoneX ? 166 : 110
    }
  })], 1);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-21f63de4_0", {
    source: ".component__tabsBar .ButtonX[data-v-21f63de4]{background-color:#fff;width:100%;box-shadow:0 .01rem 0 0 #f2f2f2;border-top:.01rem solid #eee}.component__tabsBar .ButtonX .item[data-v-21f63de4]{width:7.5rem;height:1.1rem}.component__tabsBar .ButtonX .item__btn[data-v-21f63de4]{position:relative}.component__tabsBar .ButtonX .item__btn .badge-box[data-v-21f63de4]{position:absolute;top:.02rem;left:.38rem}.component__tabsBar .ButtonX .item__btn-icon[data-v-21f63de4]{min-width:.5rem;height:.5rem;display:block}.component__tabsBar .ButtonX .item__btn-text[data-v-21f63de4]{font-size:.24rem;font-weight:400;color:#858fa2;margin-top:.05rem}.component__tabsBar .ButtonX .item__btn-text.active[data-v-21f63de4]{color:#ea313e}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-21f63de4";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
