// @wanwu/vue-nine-h5-SingleBanner/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('core-js/modules/es6.function.name');
require('core-js/modules/es6.string.link');
require('core-js/modules/es6.number.constructor');
var vueRuntimeHelpers = require('vue-runtime-helpers');

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script = {
  name: 'SingleBanner',
  props: {
    linkImg: {
      type: Object,
      "default": function _default() {
        return {};
      },
      desc: '图片'
    },
    image: {
      type: String,
      "default": ''
    },
    link: {
      type: String,
      "default": ''
    },
    height: {
      type: [Number, String],
      desc: '图片高度',
      "default": ''
    },
    marginTop: {
      type: [Number, String],
      desc: '上边距',
      "default": ''
    },
    marginBottom: {
      type: [Number, String],
      desc: '下边距',
      "default": ''
    }
  },
  computed: {
    imageStyle: function imageStyle() {
      var style = {};

      if (typeof this.height === 'number') {
        style.height = this.height / window.remRadio + 'rem';
      }

      return style;
    },
    componentStyle: function componentStyle() {
      var style = {};

      if (typeof this.marginTop === 'number') {
        style.marginTop = this.marginTop / window.remRadio + 'rem';
      }

      if (typeof this.marginBottom === 'number') {
        style.marginBottom = this.marginBottom / window.remRadio + 'rem';
      }

      return style;
    }
  },
  methods: {
    handleImageClick: function handleImageClick() {
      if (this.linkImg && this.linkImg.link || this.link) this.$location.push(this.linkImg.link || this.link);
    }
  }
};

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function() {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "div",
    {
      staticClass: "component__SingleBanner SingleBanner",
      style: _vm.componentStyle
    },
    [
      _c("img", {
        staticClass: "image",
        style: _vm.imageStyle,
        attrs: { src: (_vm.linkImg && _vm.linkImg.url) || _vm.image, alt: "" },
        on: { click: _vm.handleImageClick }
      })
    ]
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-22c52dd2_0", { source: ".component__SingleBanner[data-v-22c52dd2] {\n  width: 7.5rem;\n}\n.component__SingleBanner .image[data-v-22c52dd2] {\n  width: 100%;\n  display: block;\n}", map: {"version":3,"sources":["<input css 1>"],"names":[],"mappings":"AAAA;EACE,aAAa;AACf;AAEA;EACE,WAAW;EACX,cAAc;AAChB","file":"index.vue","sourcesContent":[".component__SingleBanner {\n  width: 7.5rem;\n}\n\n.component__SingleBanner .image {\n  width: 100%;\n  display: block;\n}"]}, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-22c52dd2";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = vueRuntimeHelpers.normalizeComponent(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    vueRuntimeHelpers.createInjector,
    undefined,
    undefined
  );

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

exports.default = __vue_component__;
