// @wanwu/base-logger/lib.es/index.js
import { fn, ua } from '@wanwu/base-fn';
import logP from './logP';
import logE from './logE';
import logR from './logR';
import VueLog from './plugins/VueLog';
import rtp from './rtp/rtp';
import config from './config';
import collect from "./util/collect";
import send from "./util/send";
import helper from "./util/helper";
import routeUtil from "./util/routeParams";
import crossUtil from "./util/cross";
import VueCreateRtp from './plugins/VueCreateRtp';
/**
 * 为指定url构造一个带rtp参数的url, 用于js代码跳转页面时，添加rtp信息
 * @param url
 * @return {string ｜ object}
 */
export function createRtpParams(url) {
    // 跳转保留最后停留时间
    rtp.jumpPageSetPageStayTime();
    if (typeof url === 'undefined') {
        return '';
    }
    // 处理url
    url = routeUtil.parseUrl(url);
    // // 获取rtp信息C、D字段为0
    // const c = 0;
    // const d = 0;
    // const rtpCnt = rtp.getRtpCnt(c, d);
    // if (url instanceof Object) {
    //     return rtp.makeObject(url, rtpCnt);
    // } else {
    //     return rtp.makeUrl(url, rtpCnt);
    // }
    // listShow agentTraceInfo_如果有trackId就listClick跳转将traceId链路一直带下去
    let traceId = sessionStorage.getItem('logger_listClickTrace') || '';
    if (!ua.isApp()) {
        let pageParam = {};
        if (location && location.search) {
            let params = location.search.split("?");
            if (params && params.length > 0) {
                pageParam = fn.parse(params[1]);
            }
        }
        if (!traceId) {
            // 如果缓存和url带的不一致那么就更新traceId traceIdSaveFlag 如果跨域需要从参数这边取值需要保存到sessionStorage
            if (pageParam.hasOwnProperty("listClickTrace") && pageParam["listClickTrace"] && pageParam['traceIdSaveFlag'] === 1) {
                sessionStorage.setItem('logger_listClickTrace', pageParam["listClickTrace"]);
                url = rtp.makeObjectTraceId(url, pageParam["listClickTrace"]);
            }
        }
        else {
            // 如果跨域给一个标识需要保存traceId 如果域名不同就再次跳转是要把traceId重新保存一下
            if (url && Object.prototype.toString.call(url) === "[object Object]" && url.origin !== location.origin) {
                url = rtp.makeObjectTraceIdSaveFlag(url, 1);
            }
            // 如果跨域给一个标识需要保存traceId
            if (url && Object.prototype.toString.call(url) === "[object String]" && url.indexOf(location.origin) === -1) {
                url = rtp.makeObjectTraceIdSaveFlag(url, 1);
            }
            url = rtp.makeObjectTraceId(url, traceId);
        }
    }
    const rtpRefer = rtp.getRtpRefer();
    if (rtpRefer) {
        if (url instanceof Object) {
            return rtp.makeObject(url, rtpRefer);
        }
        else {
            return rtp.makeUrl(url, rtpRefer);
        }
    }
    return url;
}
/**
 * 对分享页面的url进行处理，去除当前分享url的rtp参数，添加shareUserId, shareTime, firstVisit
 * 如当前url包含 shareUserId, shareTime， 则进行替换
 * @param url
 * @return {string}
 */
export function createShareParams(url) {
    if (typeof url === 'undefined') {
        return '';
    }
    const shareUserId = collect.getUserInfo().userId;
    const shareTime = collect.getTime();
    const params = {
        shareUserId,
        shareTime,
        firstVisit: 0
    };
    const shareUrl = fn.stringifyUrl(url, params);
    //return fn.removeQueryString(shareUrl, 'rtp');
    return fn.removeQueryString(shareUrl, 'rtpRefer');
}
/**
 * 对与客户端交互的url进行处理，添加当前页面的rtpCnt作为客户端的rtpUrl, 添加当前页的url为客户端的refer
 * @param {string} path
 * @param {object} query
 * @return {object}
 */
export function addReferRtp(path, query) {
    if (query instanceof Object) {
        query.refer = path;
        // 将rtpRefer添加到客户端app、小程序
        query.rtpUrl = rtp.getHashCntOrRefer(rtp.getRtpRefer());
        //query.rtpUrl = localStorage.getItem('rtpUrl');
    }
    return query;
}
/**
 * 监听页面卸载后是否还有未发送的数据
 */
// window.addEventListener('beforeunload', function() {
//     const logs = send.getLog();
//     const newLogs = send.getNewLog();
//     if (logs.length > 0)  {
//         send.sendLog(logs);
//     }
//     if (newLogs.length > 0) {
//         send.sendNewLog(logs);
//     }
// });
/**
 * 解决跨域打点通讯
 */
function unloadFn() {
    const logs = send.getLog();
    const newLogs = send.getNewLog();
    if (logs.length > 0 || newLogs.length > 0) {
        let obj = {};
        obj = JSON.parse(crossUtil.getWindowName());
        obj["logsList"] = logs.concat(newLogs) || [];
        crossUtil.setWindowName(obj);
        send.clearLog();
        send.clearNewLog();
    }
}
if (ua.isApp()) {
    // app内
    window.addEventListener('beforeunload', unloadFn);
}
else {
    // 微信浏览器没有beforeunload，ios为pagehide，安卓为unload
    if (ua.isWechat() && ua.isIOS()) {
        window.addEventListener('pagehide', unloadFn);
    }
    else {
        window.addEventListener('unload', unloadFn);
    }
}
window.addEventListener('load', function () {
    let obj = {};
    obj = JSON.parse(crossUtil.getWindowName());
    const prePageLogs = obj["logsList"] || [];
    if (prePageLogs.length > 0) {
        const logs = [];
        const newLogs = [];
        prePageLogs.forEach((log) => {
            if (log.event) {
                newLogs.push(log);
            }
            else {
                logs.push(log);
            }
        });
        // tslint:disable-next-line:no-unused-expression
        logs.length > 0 && send.sendLog(logs);
        // tslint:disable-next-line:no-unused-expression
        newLogs.length > 0 && send.sendNewLog(newLogs);
        obj["logsList"] = [];
        crossUtil.setWindowName(obj);
    }
});
// 10011 可视区域内的曝光打点
export function logE10011(elm, data) {
    if (helper.checkVisible(elm)) {
        logE(10011, data);
    }
}
const { logImmediately } = send;
// 暴露的方法
export { config, logP, logE, logR, VueLog, crossUtil, VueCreateRtp, logImmediately, };
helper.mergeParams();
send.startLogTimer();
// 设置uuid
helper.setUuid();
