// @wanwu/base-fn/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WANWUConfig = exports.md5 = exports.Ticker = exports.createTicker = exports.ImageUtils = exports.ua = exports.fn = exports.EventEmitter = exports.Semaphore = void 0;
/**
 * Create by changsheng on 2019-03-27 11:09
 */
require("./polyfill/Date");
var LocalSemaphore_1 = require("./semaphore/LocalSemaphore");
exports.Semaphore = LocalSemaphore_1.default;
var EventEmitter_1 = require("./event/EventEmitter");
exports.EventEmitter = EventEmitter_1.default;
var fn = require("./fn");
exports.fn = fn;
var ImageUtils = require("./image/ImageUtils");
exports.ImageUtils = ImageUtils;
var Ticker_1 = require("./ticker/Ticker");
Object.defineProperty(exports, "createTicker", { enumerable: true, get: function () { return Ticker_1.createTicker; } });
Object.defineProperty(exports, "Ticker", { enumerable: true, get: function () { return Ticker_1.Ticker; } });
var ua = require("./fn/ua");
exports.ua = ua;
var md5_1 = require("./crypto/md5");
exports.md5 = md5_1.default;
var config_1 = require("./config");
exports.WANWUConfig = config_1.default;
