// @wanwu/base-vue-loading-page/lib/index.es.js
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.object.assign';
import _classCallCheck from '@babel/runtime/helpers/esm/classCallCheck';
import _createClass from '@babel/runtime/helpers/esm/createClass';
import 'core-js/modules/es.number.constructor';
import loading from '@wanwu/base-vue-loading';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';
import Vue from 'vue';
import getCompDefaultOptions from '@wanwu/get-comp-default-options';

var script = {
  name: 'LoadingPage',
  components: {
    loading: loading
  },
  props: {
    size: {
      type: Number,
      "default": 100
    },
    loadingText: {
      type: String,
      "default": ''
    },
    value: {
      type: Boolean,
      "default": true
    },
    bgColor: {
      type: String,
      "default": '#f7f7f7'
    },
    loadingProps: {
      type: Object
    },
    loadType: {
      type: String,
      "default": 'line'
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _vm.value ? _c('div', {
    staticClass: "component__loadingPage loadingPage flex f-fd-c f-ai-c f-jc-c",
    style: {
      backgroundColor: _vm.bgColor
    }
  }, [_c('loading', _vm._b({
    staticClass: "loading",
    attrs: {
      "size": _vm.size,
      "load-type": _vm.loadType
    }
  }, 'loading', _vm.loadingProps, false)), _vm._v(" "), _c('div', {
    staticClass: "text"
  }, [_vm._v("\n        " + _vm._s(_vm.loadingText) + "\n    ")])], 1) : _vm._e();
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-bbe9e2be_0", {
    source: ".component__loadingPage[data-v-bbe9e2be]{position:fixed;top:0;left:0;width:100%;height:100%;z-index:1999}.component__loadingPage .text[data-v-bbe9e2be]{max-width:5.7rem;text-align:center;color:#c9c9c9;margin-top:.2rem;font-size:.26rem}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-bbe9e2be";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

var LoadingPageConstructor = Vue.extend(__vue_component__);
var instance;

var _closeFn;

var closeMixin = {
  destroyed: function destroyed() {
    _closeFn && _closeFn();
  },
  deactivated: function deactivated() {
    _closeFn && _closeFn();
  }
};
var defaultOptions = getCompDefaultOptions(__vue_component__); //  {
//     size: 100,
//     loadingText: '',
//     bgColor: '#f7f7f7',
// };

var LoadingPage = /*#__PURE__*/function () {
  function LoadingPage() {
    _classCallCheck(this, LoadingPage);
  }

  _createClass(LoadingPage, null, [{
    key: "show",
    value: function show(options) {
      if (!instance) {
        instance = new LoadingPageConstructor().$mount(document.createElement('div'));
        document.body.appendChild(instance.$el);
        instance.$on('input', function (val) {
          instance.value = val;
        });
      }

      Object.assign(instance, defaultOptions);

      if (options) {
        if (typeof options === 'string') {
          // instance.loadingText = options;
          Object.assign(instance, {
            loadingText: options
          });
        } else {
          Object.assign(instance, options);
        }
      }

      instance.value = true;
    }
  }, {
    key: "close",
    value: function close() {
      if (instance && instance.value) {
        instance.value = false;
      }

      _closeFn && _closeFn();
    }
  }]);

  return LoadingPage;
}();

__vue_component__.install = function (Vue, _ref) {
  var closeFn = _ref.closeFn;
  __vue_component__.mixin = __vue_component__.mixin || [];
  __vue_component__.mixin.push(closeMixin);
  Vue.component(__vue_component__.name, __vue_component__);
  Vue.prototype.$loadingPage = LoadingPage;
  Vue.loadingPage = LoadingPage;
  _closeFn = closeFn;
};

export default __vue_component__;
export { LoadingPage };
