// @wanwu/base-vue-load-dialog/src/index.js
import comp from './index.vue';
import Vue from 'vue';
import {
    vmBindEvents, vmUnbindEvents,
} from '@wanwu/event-utils';
const Constructor = Vue.extend(comp);
let instance;
let cacheEvents;
const defaultOptions = {
    src: '',
    showClose: true,
    jumpUrl: '',
    closeOnClickOverlay: true,
    onImgClick: null,
};
class LoadDialog {
    static show(options) {
        if (!instance) {
            instance = new Constructor().$mount();
            document.body.appendChild(instance.$el);
            instance.$on('input', (val) => {
                instance.value = val;
            });
            instance.$on('closed', () => {
                vmUnbindEvents(instance, cacheEvents);
                cacheEvents = null;
                if (instance.removeFlag) {
                    this.remove();
                }
            });
        }
        instance.removeFlag = true;
        Object.assign(instance, defaultOptions);
        if (options) {
            if (typeof options === 'string') {
                Object.assign(instance, {
                    src: options,
                });
            } else if (options.props) {
                Object.assign(instance, options.props);
            } else {
                Object.assign(instance, options);
            }

            if (options.events) {
                cacheEvents = options.events;
                vmBindEvents(instance, cacheEvents);
            }
        }
        // if (typeof options.onImgClick === 'function') {
        //     instance.$on('imgClick', options.onImgClick);
        // }
        instance.value = true;
    }

    static close() {
        if (instance && instance.value) {
            instance.value = false;
        }
    }

    static remove() {
        if (instance) {
            const parent = instance.$el.parentNode;
            if (parent) {
                parent.removeChild(instance.$el);
            }

            instance.$destroy();
        }
        instance = null;
    }
}

comp.install = (Vue) => {
    Vue.component(comp.name, comp);
    Vue.prototype.$loadDialog = LoadDialog;
    Vue.loadDialog = LoadDialog;
};
export {
    LoadDialog,
};
export default comp;
