// @wanwu/add-app-header/lib/index.es.js
import 'core-js/modules/es.array.for-each';
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.object.assign';
import 'core-js/modules/es.regexp.constructor';
import 'core-js/modules/es.regexp.exec';
import 'core-js/modules/es.regexp.to-string';
import 'core-js/modules/web.dom-collections.for-each';
import 'core-js/modules/es.array.slice';
import 'core-js/modules/es.object.define-property';
import Vue from 'vue';
import 'core-js/modules/es.function.bind';
import 'core-js/modules/es.number.constructor';
import 'core-js/modules/es.string.match';
import 'regenerator-runtime/runtime';
import _asyncToGenerator from '@babel/runtime/helpers/esm/asyncToGenerator';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import hdp from '@wanwu/hdp';
import { ua, fn } from '@wanwu/base-fn';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';
import { addThroughParams } from '@wanwu/location-plugin-through-params';
import { vmUnbindEvents, vmBindEvents } from '@wanwu/event-utils';
import getCompDefaultOptions from '@wanwu/get-comp-default-options';

(function (arr) {
  arr.forEach(function (item) {
    // eslint-disable-next-line no-prototype-builtins
    if (item.hasOwnProperty('prepend')) {
      return;
    }

    Object.defineProperty(item, 'prepend', {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function prepend() {
        var argArr = Array.prototype.slice.call(arguments);
        var docFrag = document.createDocumentFragment();
        argArr.forEach(function (argItem) {
          var isNode = argItem instanceof Node;
          docFrag.appendChild(isNode ? argItem : document.createTextNode(String(argItem)));
        });
        this.insertBefore(docFrag, this.firstChild);
      }
    });
  });
})([Element.prototype, Document.prototype, DocumentFragment.prototype]);

/*
 * @Author: 锦阳
 * @Create: 2020年12月02日
 */

/** 是否是开发模式  */

var isDevelopment = process.env.NODE_ENV === 'development';
/** 是否是在app内 */

var isAPP = ua.getAppName() === 'wwdz_b2b' || fn.isLargerThanAppVersion('2.7.7');
/** 是否在app的弹窗内 */

var isEmbed = fn.getQueryString('isEmbed');
/** 是否需要展示app头部 */

var needShowAppHeader = (isAPP || isDevelopment) && !isEmbed;

var _listener = function _listener() {};

if (needShowAppHeader) {
  var title = document.querySelector('head>title'); // 选择需要观察变动的节点

  var targetNode = title; // 观察器的配置（需要观察什么变动）

  var config = {
    attributes: true,
    childList: true,
    subtree: true
  }; // 当观察到变动时执行的回调函数

  var callback = function callback() {
    // Use traditional 'for loops' for IE 11
    if (typeof _listener === 'function') {
      _listener();
    }
  }; // 创建一个观察器实例并传入回调函数


  var observer = new MutationObserver(callback); // 以上述配置开始观察目标节点

  observer.observe(targetNode, config);
}
/**
 *  监听title改动
 * @param {(title:string)=>void} handler 监听函数
 * @param {object} options 监听配置选项
 * @param {boolean} options.immediate 是否在初始化监听时触发
 */


function watchDocumentTitle(handler, _ref) {
  var immediate = _ref.immediate;
  typeof handler === 'function' && immediate && handler();
  _listener = handler;
}

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
var script = {
  name: 'AppHeader',
  props: {
    // 默认占位，实现原生APP内的头部效果
    // 针对店铺首页shop/index进行absolute定位处理
    top: {
      type: [Number, String]
    },
    // 是否自定义 -- 1.是：无背景，无nh，使用自定义top 2. 否：有背景，有nh，不使用自定义top
    custom: {
      type: Boolean,
      "default": false
    },
    // 是否展示documentTitle
    showTitle: {
      type: Boolean,
      "default": true
    },
    // title 样式更改
    titleColor: {
      type: String
    },
    headerStyle: {
      type: Object
    },
    // 是否展示分享按钮
    showShare: {
      type: Boolean,
      "default": false
    },
    // 组件使用的header隐藏显示
    showAppHeader: {
      type: Boolean,
      "default": false
    },
    // 自定义头部背景
    topBg: {
      type: String,
      "default": ''
    },
    // nh高度背景
    nhBgColor: {
      type: String,
      "default": '#fff'
    },
    // 是否默认展示（九宫预渲染专用）
    initShow: Boolean,
    // 按钮配置
    headerBtn: {
      type: Object,
      "default": function _default() {
        return {
          back: {
            // btnBlackImgUrl: 'https://cdn.wanwudezhi.com/static/web-static/image/41aa6ba94c521e5159a96f6ffd0d36c6_72x72.png',
            // btnTransparentImgUrl: 'https://cdn.wanwudezhi.com/static/web-static/image/1bcff5b9e19f3a3c1ea173719d0da18b_72x72.png',
            icon: 'ww-icon-fanhui5',
            show: true
          },
          title: {
            title: document.title,
            show: true
          },
          refresh: {
            // btnBlackImgUrl: 'https://cdn.wanwudezhi.com/static/web-static/image/768d6767efad5a439b94de30b34dcd7e_75x72.png',
            // btnTransparentImgUrl: 'https://cdn.wanwudezhi.com/static/web-static/image/39fe7fe465725718632d61275bb8f8da_75x72.png',
            icon: 'ww-icon-shuaxin1',
            show: true
          },
          share: {
            // btnTransparentImgUrl: 'https://cdn.wanwudezhi.com/static/web-static/image/630986d28d520a4682e0bc823c9f98a5_72x72.png',
            // btnBlackImgUrl: 'https://cdn.wanwudezhi.com/static/web-static/image/63a786e8f9737d402f999928235e88b3_72x72.png',
            icon: 'ww-icon-fenxiang1',
            show: false
          }
        };
      }
    }
  },
  data: function data() {
    return {
      hasDomain: false,
      nh: null,
      showHeader: this.showAppHeader || false,
      documentTitle: '玩物得志商城',
      noShare: true,
      appHeaderHeight: 0,
      uaNh: 0,
      isDP: !!this.getUaNhFromReg('StatusBarHeightDP')
    };
  },
  computed: {
    _headerStyle: function _headerStyle() {
      var isImg = /^http(s)?:\/\//.test(this.topBg);
      var bgObj = isImg ? {
        backgroundImage: this.topBg ? "url(".concat(this.topBg, ")") : '',
        backgroundSize: '100%',
        backgroundRepeat: 'no-repeat',
        backgroundPosition: 'center 99%'
      } : {
        backgroundColor: this.topBg
      };
      return _objectSpread({
        height: this.appHeaderHeight + 'rem',
        top: this.top && this.custom ? this.top + 'rem' : 0
      }, bgObj);
    },
    titleStyle: function titleStyle() {
      return {
        color: this.titleColor
      };
    },
    nhStyle: function nhStyle() {
      return {
        height: this.nh + 'rem',
        background: this.topBg ? 'transparent' : this.nhBgColor
      };
    },
    isWhiteBg: function isWhiteBg() {
      return !(this.custom || this.topBg);
    }
  },
  watch: {
    showAppHeader: function showAppHeader(newVal) {
      if (newVal) {
        this.setPublicHeaderHeight();
      }
    }
  },
  created: function created() {
    this.setDocumentTitle = this.setDocumentTitle.bind(this);
    this.switchAppHeader();
    this.getUaNh();
  },
  activated: function activated() {
    this.switchAppHeader();
  },
  methods: {
    getUaNh: function getUaNh() {
      this.uaNh = this.getUaNhFromReg('StatusBarHeightDP') * 2 || this.getUaNhFromReg('StatusBarHeight');
    },
    getUaNhFromReg: function getUaNhFromReg(reg) {
      var uaNhReg = new RegExp("\\s".concat(reg, "/([0-9]+)\\s?"));
      var uaNhMetch;

      try {
        uaNhMetch = navigator.userAgent.match(uaNhReg);
      } catch (error) {
        console.error(error);
      }

      return Number(uaNhMetch && uaNhMetch[1]);
    },
    // 开关
    switchAppHeader: function switchAppHeader() {
      var _this = this;

      if (needShowAppHeader || this.initShow) {
        // this.show();
        this.$nextTick(function () {
          return watchDocumentTitle(_this.setDocumentTitle, {
            immediate: true
          });
        });
        this.setPublicHeaderHeight();

        if (this.showShare) {
          this.showAppHeaderShare();
        } else {
          this.hideAppHeaderShare();
        }
      } else {
        this.hide();
      }
    },
    // 提供给外部使用的隐藏分享的功能
    hideAppHeaderShare: function hideAppHeaderShare() {
      this.noShare = true;
      this.$set(this.headerBtn['share'], 'show', false);
    },
    showAppHeaderShare: function showAppHeaderShare() {
      this.noShare = false;
      this.$set(this.headerBtn['share'], 'show', true);
    },
    // 提供给外部使用的更改头部背景的方法
    setTopBg: function setTopBg(url) {
      this.topBg = url;
    },
    setTitleColor: function setTitleColor(val) {
      this.titleColor = val;
    },
    setNhBgColor: function setNhBgColor(val) {
      this.nhBgColor = val;
    },
    setDocumentTitle: function setDocumentTitle() {
      this.$set(this.headerBtn['title'], 'title', document.title);
    },
    // 自定义Header包含功能：后退，刷新，分享
    clickFn: function clickFn(btnFn) {
      this.$emit(btnFn);

      if (btnFn === 'back') {
        ua.isApp() ? hdp.execRequest('PluginNavigation', 'pageFinish') : this.$location.back();
      }

      if (btnFn === 'refresh') {
        location.reload();
      }

      if (btnFn === 'share' && typeof window.commonShareClick === 'function') {
        window.commonShareClick();
      }
    },
    // 设置拥有appHeader的document body的样式
    setDocumentBodyStyle: function setDocumentBodyStyle() {
      document.body.style.paddingTop = this.appHeaderHeight + 'rem';
    },
    // 设置没有appHeader的document body的样式
    resetDocumentBodyStyle: function resetDocumentBodyStyle() {
      window.appHeaderHeight = 0;
      this.appHeaderHeight = 0;
      window.appHeaderNH = 0;
      document.body.style.paddingTop = '0';
    },
    // 设置提供给全局使用的appHeaderHeight与nh（手机状态栏高度）
    setPublicHeaderHeight: function () {
      var _setPublicHeaderHeight = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var nh, response;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                // 获取各个手机的状态栏
                nh = fn.getQueryString('__nhDP') * 2 || fn.getQueryString('__nh');
                this.getUaNh();

                if (!(!nh && isAPP)) {
                  _context.next = 14;
                  break;
                }

                _context.prev = 3;
                _context.next = 6;
                return hdp.execRequest('PluginNavigation', 'pageStatusHeight');

              case 6:
                response = _context.sent;

                if (response) {
                  nh = response.__nhDP * 2 || response.__nh;
                  if (!this.uaNh) this.setDocumentBodyStyle();
                } else {
                  nh = 0;
                }

                _context.next = 14;
                break;

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](3);
                console.log(_context.t0);
                nh = 0;

              case 14:
                this.setNh(nh);

              case 15:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[3, 10]]);
      }));

      function setPublicHeaderHeight() {
        return _setPublicHeaderHeight.apply(this, arguments);
      }

      return setPublicHeaderHeight;
    }(),
    // 兼容异步获取nh
    setNh: function setNh() {
      var nh = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
      this.getUaNh();
      var NumberNh = Number(nh);

      if (ua.getAppName() === 'wwdz_b2b' || fn.isLargerThanAppVersion('3.1.8')) {
        if (ua.isIOSApp()) {
          NumberNh = NumberNh * 2;
        }
      } else {
        if (ua.isIOSApp()) {
          NumberNh = NumberNh + 10;
        } else if (ua.isAndroidApp() && Number(nh)) {
          NumberNh = NumberNh - 20;
        }
      }

      if (!ua.isIOSApp()) {
        NumberNh = this.uaNh || NumberNh;
      }

      var appHeaderHeight = (88 + NumberNh) / 100;
      var appHeaderNH = NumberNh / 100;
      this.nh = appHeaderNH; // 自定义头部高度

      window.appHeaderHeight = appHeaderHeight;
      this.appHeaderHeight = appHeaderHeight;
      window.appHeaderNH = appHeaderNH;
    },
    // 隐藏
    hide: function hide() {
      this.resetDocumentBodyStyle();
      this.showHeader = false;
    },
    // 展示
    show: function show() {
      watchDocumentTitle(this.setDocumentTitle, {
        immediate: true
      });
      this.resetDocumentBodyStyle();
      this.setPublicHeaderHeight();
      this.setDocumentBodyStyle();
      this.showHeader = true;
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _vm.showHeader ? _c('div', {
    staticClass: "component__AppHeader appHeader bg",
    "class": {
      'custom': _vm.custom,
      white: !_vm.isWhiteBg
    },
    style: _vm.headerStyle || _vm._headerStyle
  }, [_vm.nh && !_vm.custom ? _c('div', {
    staticClass: "appHeader-nh",
    style: _vm.nhStyle
  }) : _vm._e(), _vm._v(" "), _c('div', {
    staticClass: "appHeader-wrapper flex f-ai-c f-jc-c"
  }, [_c('div', {
    staticClass: "left_btn absolute flex f-ai-c"
  }, [_vm._t("back", [_vm.headerBtn.back.show ? [(!_vm.isWhiteBg ? _vm.headerBtn.back.btnTransparentImgUrl : _vm.headerBtn.back.btnBlackImgUrl) ? _c('img', {
    staticClass: "back",
    attrs: {
      "src": !_vm.isWhiteBg ? _vm.headerBtn.back.btnTransparentImgUrl : _vm.headerBtn.back.btnBlackImgUrl
    },
    on: {
      "click": function click($event) {
        return _vm.clickFn('back');
      }
    }
  }) : _vm.headerBtn.back.icon ? _c('i', {
    staticClass: "WWDZ back icon",
    "class": [_vm.headerBtn.back.icon],
    on: {
      "click": function click($event) {
        return _vm.clickFn('back');
      }
    }
  }) : _vm._e()] : _vm._e()], null, _vm.headerBtn.back)], 2), _vm._v(" "), _vm._t("title", [_vm.headerBtn.title.show && _vm.showTitle ? _c('span', {
    staticClass: "title",
    style: _vm.titleStyle,
    domProps: {
      "textContent": _vm._s(_vm.headerBtn.title.title)
    }
  }) : _vm._e()], null, _vm.headerBtn.title), _vm._v(" "), _c('div', {
    staticClass: "right_btn flex f-ai-c absolute"
  }, [_vm._l(['refresh', 'share'], function (key) {
    return [_vm._t(key, [_vm.headerBtn[key].show ? [(!_vm.isWhiteBg ? _vm.headerBtn[key].btnTransparentImgUrl : _vm.headerBtn[key].btnBlackImgUrl) ? _c('img', {
      key: key,
      "class": [key],
      attrs: {
        "src": !_vm.isWhiteBg ? _vm.headerBtn[key].btnTransparentImgUrl : _vm.headerBtn[key].btnBlackImgUrl
      },
      on: {
        "click": function click($event) {
          return _vm.clickFn(key);
        }
      }
    }) : _vm.headerBtn[key].icon ? _c('i', {
      key: key,
      staticClass: "WWDZ icon",
      "class": [_vm.headerBtn[key].icon, key],
      on: {
        "click": function click($event) {
          return _vm.clickFn(key);
        }
      }
    }) : _vm._e()] : _vm._e()], null, _vm.headerBtn[key])];
  })], 2)], 2)]) : _vm._e();
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-9e704034_0", {
    source: ".appHeader[data-v-9e704034]{position:fixed;top:0;left:0;width:100%;z-index:3000;color:#2e333b}.appHeader-nh[data-v-9e704034]{width:100%;background:#fff}.appHeader.white[data-v-9e704034]{color:#fff}.appHeader-wrapper[data-v-9e704034]{height:.88rem;padding:.2rem 0}.appHeader-wrapper .icon[data-v-9e704034]{font-size:.48rem}.appHeader-wrapper .title[data-v-9e704034]{font-size:.36rem;white-space:nowrap;max-width:4.6rem;overflow:hidden;text-overflow:ellipsis}.appHeader-wrapper .share[data-v-9e704034]{margin-left:.48rem}.appHeader-wrapper img[data-v-9e704034]{width:.48rem;display:block}.appHeader-wrapper .absolute[data-v-9e704034]{position:absolute}.appHeader-wrapper .absolute.left_btn[data-v-9e704034]{left:.28rem}.appHeader-wrapper .absolute.right_btn[data-v-9e704034]{right:.28rem}.bg[data-v-9e704034]{background:#fff}.custom.bg[data-v-9e704034]{background:0 0}.custom.bg>div[data-v-9e704034]:nth-child(2){color:#fff}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-9e704034";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = /*#__PURE__*/normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

var AppHeaderConstructor = Vue.extend(__vue_component__); // APP内H5自定义头部使用的nh(设备状态栏高度)

addThroughParams('__nh');
var cacheEvents;
var defaultOptions = getCompDefaultOptions(__vue_component__);
/** 全局appHeader实例 */

var instance;
/**
 * 添加头部
 * @param {object} param 参数
 * @param {string} param.path 路径（一般是当前路径）
 * @param {string[]} param.hasSharePath 需要展示分享按钮的路径规则正则字符串数组
 * @param {string[]} param.ignoreHeaderPath 不需要展示头部的路径规则正则字符串数组
 * @param {object} param.props addHeader props属性
 * @param {object} param.events addHeader 事件
 * @param {object} param.slots addHeader 插槽 @example {back:()=>(<a>返回</a>)}
 * @returns {Vue}
 */

function addAppHeader(_ref) {
  var _ref$path = _ref.path,
      path = _ref$path === void 0 ? '' : _ref$path,
      _ref$hasSharePath = _ref.hasSharePath,
      hasSharePath = _ref$hasSharePath === void 0 ? [] : _ref$hasSharePath,
      _ref$ignoreHeaderPath = _ref.ignoreHeaderPath,
      ignoreHeaderPath = _ref$ignoreHeaderPath === void 0 ? [] : _ref$ignoreHeaderPath,
      props = _ref.props,
      events = _ref.events,
      slots = _ref.slots;

  if (needShowAppHeader) {
    if (!instance) {
      instance = new AppHeaderConstructor();
      instance.vm = instance.$mount();
      document.body.prepend(instance.vm.$el);
    }

    vmUnbindEvents(instance, cacheEvents);
    Object.assign(instance.$props, defaultOptions);

    if (props) {
      Object.assign(instance.$props, props);
    }

    if (events) {
      cacheEvents = events;
      vmBindEvents(instance, cacheEvents);
    }

    instance.$scopedSlots = {};

    if (slots) {
      Object.assign(instance.$scopedSlots, slots);
    } // 隐藏头部


    var isIgnoreHeader = checkHasPath(ignoreHeaderPath, path); // 拥有分享的头部

    var isHasShareHeader = checkHasPath(hasSharePath, path);

    if (isIgnoreHeader) {
      instance.hide();
    } else {
      instance.show();
      instance.setPublicHeaderHeight();

      if (isHasShareHeader) {
        instance.showAppHeaderShare();
      } else {
        instance.hideAppHeaderShare();
      }
    }
  }

  return instance;
}
/**
 * 公用校验路径的方法
 * @param {string[]|RegExp[]} pathConfigArray 路径规则（正则字符串）数组
 * @param {string} path 待校验的路径字符串
 */


function checkHasPath(pathConfigArray, path) {
  var status = false;
  pathConfigArray.forEach(function (item) {
    if (new RegExp(item).test(path)) {
      status = true;
    }
  });
  return status;
}

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export { __vue_component__ as AppHeader, addAppHeader, instance as appHeader, checkHasPath, isAPP, isDevelopment, isEmbed, needShowAppHeader };
