// @wanwu/base-share/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('core-js/modules/es.array.concat');
require('core-js/modules/es.array.for-each');
require('core-js/modules/es.array.iterator');
require('core-js/modules/es.array.join');
require('core-js/modules/es.array.map');
require('core-js/modules/es.date.now');
require('core-js/modules/es.date.to-string');
require('core-js/modules/es.object.to-string');
require('core-js/modules/es.parse-int');
require('core-js/modules/es.regexp.constructor');
require('core-js/modules/es.regexp.exec');
require('core-js/modules/es.regexp.to-string');
require('core-js/modules/es.string.iterator');
require('core-js/modules/es.string.replace');
require('core-js/modules/es.string.split');
require('core-js/modules/es.string.link');
require('core-js/modules/web.dom-collections.iterator');
require('core-js/modules/web.url');
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/esm/defineProperty'));
var WxSdk = _interopDefault(require('@wanwu/wx-sdk'));
var baseFn = require('@wanwu/base-fn');
var baseLogger = require('@wanwu/base-logger');
var User = _interopDefault(require('@wanwu/base-user'));

var shareImages = ['https://cdn.wanwudezhi.com/static/mall-web/images/846ef73cb6af50d4a541059aee8d9994.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/715b68fda2b9b663ba47e0d926de5056.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/498beb26a827552e87e9de0c8fa2b051.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/dba5648780a1e5b683898f5f5c217300.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/49356103af6c8b6a2e416b1074e01d4e.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/82f4eea2ad3c5d8a2b6fca6768c51541.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/a75bedeb1260d97ad5abb5850d8266b0.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/5210a5e51004afcf34ec0a2d7389fdfd.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/c514c60b493ad43e295196881d221e63.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/25f1fd5db22212df7332499935df29e8.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/5c1c4964a981f7f98791b7b471cc38fb.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/cc63cf1b7b040fa4b40b1849785fdc3c.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/1ccf7fcc708426849df93e73faff3eae.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/b3885865d82e486aa22a37d60d8a4571.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/7ecf98c6d3b0fde32134792ab14ed642.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/bbe6a49ddc4fd0f9b060b0ed1cc6b39f.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/c620703f78cbc57a59cc8ba901590dcc.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/f80a96de4491d7d757b7caff9de5b49f.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/7c066517dc7c04a10286d8cac1f3d53e.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/f22bffcd69f591ee98fa08a441637078.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/f493e6c0f75f0340075f9dd67d1d8271.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/a4221f5927b4f3274737b39662adc5c7.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/e0156a39c152cb4e7865e25584538b18.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/99ff0cc4d30b967ebf988e013b6657bb.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/4abffb5401097f0920b8fecdf4bf8cd9.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/48b741279be121db9ac79400d3097506.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/509fba75aba1809eabaa1c0ffda67ba3.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/789c9bea256c1c952d4b832330c5b54e.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/d23532956caff8e8969662a2d1d93176.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/c6ade410931cb8d882da4ed8c936f0dd.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/7bdb1b63ff8055e0c155a4e16d6c9574.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/fd2e644421701b3bddac75e8a28a3220.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/1bc4e8eef67d58f9f400e17a6e5cbb4e.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/036611e8455ceb52e0d88ae8860a95cc.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/395c3640e4de528ca0a4c2986c643db9.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/79f4d5a9fc9b2d8c0bc7a5f62153151f.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/dc083e2fd98624144e3fdf729c1b55d5.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/c6260fd5c1c248975ed5074877c602b9.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/cb27c2a86a0e48073d5bb7f3e2a8da8e.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/73e6ea367e48ef9cadf5a672ad88d69d.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/78c72ffd78a454042beb0bf657fc6c38.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/468db83bbdee7bfd6e960677070f0597.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/49849f22e6fa80594eca21691cd30c5a.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/0b2d4c7a417567b123752734c5eb0c02.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/71d7001c1461b7300694a4e594dd8c9f.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/6547ad12a03c74d3a7ec44afb42b6197.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/09640071cadfaa00d8d6a047f19638e1.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/bf14f9e40ab0d2bf91363cd75edaa92a.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/888d896f6b3d7ebb79b70ddefcafecd4.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/6e52e012b5dc334a7b1dfbf430220dec.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/9935f1ca7c20351ae002d8594ba880d7.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/9dcd3892f2526f79561d491878c394e7.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/3bf3458d1f0c3b442776d5f921382bba.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/1318415e1aebc6eb8cff0ecdb23f4442.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/8aa27133084cfeeebf2e0c76e30e1de2.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/0acffa61c23866fdff517e744735b0cf.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/44ba2a24cf0dca1dc4ec79a6fb1ebcc5.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/5ab923b09e4e400543509d0cadba9e67.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/06b2fdf9fd68ac9e2335d40f6be1a5af.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/ace73e061c776b776576dde66b692f1a.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/5040581c2338d2729c3dd8e48eb63f58.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/ee95d26875884180cce22fbb45e17ebf.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/46313038615f467da309bfe64ea13ee8.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/257ab86b1f97458696830c3c61cd5292.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/644801da24286d117c9c1c61bc45ef65.png', 'https://cdn.wanwudezhi.com/static/mall-web/images/ace8fa090c9089659fc6a88cf10e918f.png'];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
var CacheShareOptions;
var CurrentRouterPath; // 当前路由

var WhiteReg = ['h5.wanwudezhi.com/mall-web', 'h5.wanwudezhi.com/product-web', 'h5.wanwudezhi.com/live-web', 'h5.wanwudezhi.com/promotion-web', 'h5.wanwudezhi.com/community-web', 'm.wanwudezhi.com', 'marketing-web'];
var FakeQueryKeys = ['shareTime'];
var ShareRemoveKeys = ['token', 'trace_info', 'tx_trace_key', 'cps_user_source'];
var index = {
  /**
   * 更新页面分享配置，缓存分享配置对象
   * @param {object} options
   * @param {string} options.title 页面分享标题
   * @param {string} options.desc 页面分享描述
   * @param {string} options.link 页面分享链接
   * @param {string} options.imgUrl 页面分享小图
   * @param {string} options.captureScreen 页面使用截屏
   */
  updateShareData: function updateShareData(options) {
    if (!baseFn.ua.isWechat()) {
      CacheShareOptions = this.getShareOptions(options);
      return;
    }

    this.showShareMenu();
    var shareOptions = this.getShareOptions(options); // 当captureScreen置为true时，不再传imgUrl

    if (options.captureScreen && baseFn.ua.isWechatMiniProgram()) {
      shareOptions.imgUrl = 'none';
    }

    CacheShareOptions = shareOptions;
    this.miniProgramPostMessage();
    WxSdk.updateAppMessageShareData(shareOptions);
    WxSdk.updateTimelineShareData(shareOptions);
    WxSdk.onMenuShareAppMessage(shareOptions);
    WxSdk.onMenuShareTimeline(shareOptions);
  },

  /**
   * 生成分享配置对象，处理link，生成分享准确的分享链接，
   * 这个方法可以认为是一个内部方法，
   * 一般不建议直接调用，可以用`getCacheShareOptions`方法
   * @param {object} options
   * @param {string} options.title 页面分享标题
   * @param {string} options.desc 页面分享描述
   * @param {string} options.link 页面分享链接
   * @param {string} options.imgUrl 页面分享小图
   * @returns {object} shareOptions 分享配置对象
   * {title,desc,link,imgUrl}
   */
  getShareOptions: function getShareOptions() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    CurrentRouterPath = options.router || CurrentRouterPath;
    var defaultShareOptions = {
      title: '玩物得志商城',
      desc: '找国风好物，上玩物得志',
      link: location.href,
      imgUrl: baseFn.ImageUtils.getSquareImage(this.getRandomShareImg(), 200),
      router: CurrentRouterPath
    };

    var _ref = options || {},
        title = _ref.title,
        desc = _ref.desc,
        link = _ref.link,
        imgUrl = _ref.imgUrl,
        target = _ref.target;

    var shareOptions = {
      title: title || defaultShareOptions.title,
      desc: desc || defaultShareOptions.desc,
      link: link || defaultShareOptions.link,
      imgUrl: imgUrl || defaultShareOptions.imgUrl,
      router: CurrentRouterPath,
      target: target || 'xcx'
    };
    var userInviteCode = User.getUserInfo().inviteCode;

    if (userInviteCode) {
      var url = shareOptions.link;
      shareOptions.link = baseFn.fn.setQueryString(url, 'inviteCode', userInviteCode);
    }

    var isPublicFlow = baseFn.fn.getQueryString('isPublicFlow');

    if (isPublicFlow) {
      var _url = shareOptions.link;
      shareOptions.link = baseFn.fn.setQueryString(_url, 'isPublicFlow', '0');
    } // 分享渠道


    var sf = baseFn.fn.getQueryString('sf', shareOptions.link);

    if (!sf) {
      var _url2 = shareOptions.link;
      shareOptions.link = baseFn.fn.setQueryString(_url2, 'sf', 'h5');
    } // const token = fn.getQueryString('token');
    // if (token) {
    //     const url = shareOptions.link;
    //     shareOptions.link = fn.removeQueryString(url, 'token');
    // }
    // const trace_info = fn.getQueryString('trace_info');
    // if (trace_info) {
    //     const url = shareOptions.link;
    //     shareOptions.link = fn.removeQueryString(url, 'trace_info');
    // }


    ShareRemoveKeys.map(function (removeKey) {
      var val = baseFn.fn.getQueryString(removeKey);
      if (!val) return;
      var url = shareOptions.link;
      shareOptions.link = baseFn.fn.removeQueryString(url, removeKey);
    });
    shareOptions.link = baseLogger.createShareParams(shareOptions.link);
    /**
     * 对符合白名单要求的链接做处理
     * link会带上当前时间戳
     * 并且生成一个realLink，不带时间戳的链接，目前是作为生成短链的缓存的key
     */

    var linkObj = new URL(shareOptions.link);

    if (new RegExp("(".concat(WhiteReg.join('|'), ")")).test(linkObj.origin + linkObj.pathname)) {
      var pattern = /(.*)\/\d+$/;

      if (pattern.test(linkObj.pathname)) {
        linkObj.pathname = linkObj.pathname.replace(pattern, '$1');
        shareOptions.realLink = linkObj.toString();
      }

      linkObj.pathname += "/".concat(Date.now());
      shareOptions.link = linkObj.toString();
    }

    shareOptions.realLink = shareOptions.realLink || shareOptions.link;
    FakeQueryKeys.forEach(function (key) {
      shareOptions.realLink = baseFn.fn.removeQueryString(shareOptions.realLink, key);
    });
    return shareOptions;
  },

  /**
   * 生成当前页面缓存的分享配置对象，或者在不影响页面级别分享配置下，生成分享配置对象
   * @param {object} options
   * @param {string} options.title 分享标题
   * @param {string} options.desc 分享描述
   * @param {string} options.link 分享链接
   * @param {string} options.imgUrl 分享小图
   * @returns {object} cacheShareOptions 缓存再处理的，或者基于新的配置生成的分享对象
   * {title,desc,link,imgUrl,`page`,`queryStr`}  [page,queryStr是专为分享小程序生成的属性]
   */
  getCacheShareOptions: function getCacheShareOptions(options) {
    var newShareOptions = options ? this.getShareOptions(options) : CacheShareOptions; // 获取微信的accessToken的值

    var originUrl = new RegExp("^".concat(location.origin, "(/mall-web)?"));
    var urlParse = newShareOptions.link.replace(originUrl, '').split('?');
    var page = urlParse[0];
    var queryStr = urlParse[1];

    if (baseFn.ua.isWechatMiniProgram()) {
      var replacePage = [{
        rule: '/shop/index',
        xcx: '/store/index'
      }];

      for (var _i = 0, _replacePage = replacePage; _i < _replacePage.length; _i++) {
        var item = _replacePage[_i];
        var rule = item.rule;

        if (!(rule instanceof RegExp)) {
          rule = new RegExp("".concat(rule, "(/|$)+"));
        }

        if (rule.test(page)) {
          page = item.xcx;
        }

        break;
      }
    }

    if (newShareOptions.target === 'h5') {
      page = '/webview/index';
      queryStr = "".concat(urlParse[1] + '&' || '', "src=").concat(encodeURIComponent(newShareOptions.link));
    }

    return _objectSpread(_objectSpread({}, newShareOptions), {}, {
      page: page,
      queryStr: queryStr
    });
  },

  /**
   * 隐藏微信分享菜单中的按钮，效果有问题，不建议使用
   */
  hideShareMenu: function hideShareMenu() {
    WxSdk.hideMenuItems({
      menuList: ['menuItem:share:appMessage', 'menuItem:share:timeline', 'menuItem:share:qq', 'menuItem:share:weiboApp', 'menuItem:share:facebook', 'menuItem:share:QZone', 'menuItem:copyUrl', 'menuItem:openWithQQBrowser', 'menuItem:openWithSafari', 'menuItem:share:email']
    });
  },

  /**
   * 显示微信分享菜单中的按钮，效果有问题，不建议使用
   */
  showShareMenu: function showShareMenu() {
    WxSdk.showMenuItems({
      menuList: ['menuItem:share:appMessage', 'menuItem:share:timeline' // 'menuItem:share:qq',
      // 'menuItem:share:weiboApp',
      // 'menuItem:share:facebook',
      // 'menuItem:share:QZone',
      // 'menuItem:copyUrl',
      // 'menuItem:openWithQQBrowser',
      // 'menuItem:openWithSafari',
      // 'menuItem:share:email',
      ]
    });
  },
  getRandomShareImg: function getRandomShareImg() {
    var randomNum = parseInt(Math.random() * 20);
    return shareImages[randomNum];
  },

  /**
   * 在用户触发分享或者其他会向小程序通信的钩子后，向小程序传输数据
   * @param {object} data 需要传给小程序的值
   */
  miniProgramPostMessage: function miniProgramPostMessage() {
    var data = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    if (baseFn.ua.isWechatMiniProgram()) {
      try {
        var shareOptions = this.getCacheShareOptions(); // shareOptions.link = `pages/webview/index?src=${encodeURIComponent(shareOptions.link)}`;

        shareOptions.link = "pages".concat(shareOptions.page, "?").concat(shareOptions.queryStr);
        var params = {
          shareOptions: shareOptions,
          businessData: data
        };
        window.wx.miniProgram.postMessage({
          data: params
        });
      } catch (e) {
        console.log(e);
      }
    }
  }
};

exports.default = index;
