// @wanwu/env-utils/lib/index.es.js
import 'core-js/modules/es.regexp.exec';
import 'core-js/modules/es.string.replace';
import { fn } from '@wanwu/base-fn';

/**
 * 获取api的env参数
 */

function getApiEnv() {
  return fn.getQueryString('__env');
}
/**
 * 获取前端的env参数
 * 不符合`/^https?:\/\/(.+)(\.|-)(h5|m)\.wanwudezhi\.com/`的域名会返回当前域名
 */

function getFeEnv() {
  return location.origin.replace(/^https?:\/\/(.+)(\.|-)(h5|m)\.wanwudezhi\.com/, "$1");
}
/**
 * 是否是后端预发
 */

function isApiPre() {
  return getApiEnv() === 'pre';
}
/**
 * 是否是后端测试
 */

function isApiDev() {
  return getApiEnv() === 'develop';
}
/**
 * 是否是后端线上
 */

function isApiProd() {
  return !isApiPre() && !isApiDev();
}
/**
 * 是否是前端预发
 */

function isFePre() {
  return getFeEnv() === 'pre';
}
/**
 * 是否是前端测试
 */

function isFeDev() {
  return getFeEnv() === 'dev';
}
/**
 * 是否是前端线上
 */

function isFeProd() {
  return !isFePre() && !isFeDev();
}
var index = {
  getApiEnv: getApiEnv,
  getFeEnv: getFeEnv,
  isApiPre: isApiPre,
  isApiDev: isApiDev,
  isApiProd: isApiProd,
  isFePre: isFePre,
  isFeDev: isFeDev,
  isFeProd: isFeProd
};

export default index;
export { getApiEnv, getFeEnv, isApiDev, isApiPre, isApiProd, isFeDev, isFePre, isFeProd };
