// @wanwu/base-vue-badge/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('core-js/modules/es.function.name');
require('core-js/modules/es.number.constructor');
require('@wanwu/base-style');
var vueRuntimeHelpers = require('@wanwu/vue-runtime-helpers');

var script = {
  name: 'Badge',
  props: {
    number: {
      type: [Number, String],
      "default": 0
    },
    maxNumber: {
      type: [Number, String],
      "default": 99
    },
    isDot: {
      type: Boolean,
      "default": false
    },
    status: {
      // success(#52c41a)、error(#EA3131)、disabled(#d9d9d9)、info(#1890ff)、warning(#faad14)
      type: String,
      "default": 'error'
    }
  },
  computed: {
    countNum: function countNum() {
      var number = this.number,
          maxNumber = this.maxNumber,
          isDot = this.isDot;
      if (isDot) return;
      return number > maxNumber ? "".concat(maxNumber, "+") : number;
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _obj;

  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    "class": (_obj = {
      'component_badge': true,
      'flex': true,
      'f-jc-c': true,
      'f-ai-c': true,
      dot: _vm.isDot,
      badge: !_vm.isDot
    }, _obj[_vm.status + "-class"] = _vm.status, _obj)
  }, [_vm._v("\n    " + _vm._s(_vm.countNum) + "\n")]);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-76dd8d18_0", {
    source: ".badge[data-v-76dd8d18]{padding:0 .08rem;height:.28rem;line-height:.28rem;font-size:.24rem;color:#fff;border-radius:.14rem}.dot[data-v-76dd8d18]{width:.16rem;height:.16rem;border-radius:100%}.disabled-class[data-v-76dd8d18]{background-color:#d9d9d9}.error-class[data-v-76dd8d18]{background-color:#ea3131}.success-class[data-v-76dd8d18]{background-color:#52c41a}.info-class[data-v-76dd8d18]{background-color:#1890ff}.warning-class[data-v-76dd8d18]{background-color:#faad14}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-76dd8d18";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = vueRuntimeHelpers.normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, vueRuntimeHelpers.createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

exports.default = __vue_component__;
