// @wanwu/base-vue-popup/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('core-js/modules/es.function.name');
require('core-js/modules/es.object.assign');
require('core-js/modules/es.string.fixed');
require('core-js/modules/es.number.constructor');
require('core-js/modules/es.parse-int');
var overlay = _interopDefault(require('@wanwu/base-vue-overlay'));
var baseFn = require('@wanwu/base-fn');
var vueRuntimeHelpers = require('@wanwu/vue-runtime-helpers');

var script = {
  name: 'Popup',
  components: {
    overlay: overlay
  },
  directives: {
    fixed: {
      inserted: function inserted(el, fixed, vnode) {
        if (!fixed.value) return;
        var scrollTop = document.body.scrollTop || document.documentElement.scrollTop;

        if (window.currentRouterView) {
          window.currentRouterView.style.overflow = 'hidden !important';
        } else {
          document.body.style.cssText += 'position:fixed;width:100%;height:100%;top:-' + scrollTop + 'px;';
        }
      },
      // unbind 指令与元素解绑时调用
      unbind: function unbind(el, fixed, vnode) {// if (fixed.value) vnode.context.unBindFixed();
      }
    }
  },
  props: {
    value: {
      type: Boolean,
      "default": false
    },
    overlay: {
      type: Boolean,
      "default": true
    },
    closeOnClickOverlay: {
      type: Boolean,
      "default": true
    },
    position: {
      type: String,
      "default": ''
    },
    overlayStyle: {
      type: [Object, String]
    },
    slotStyle: {
      type: [Object, String]
    },
    // 是否允许页面滚动
    fixed: {
      type: Boolean,
      "default": false
    },
    wapperClass: {
      type: [Object, String]
    },
    duration: {
      type: Number,
      "default": 300
    }
  },
  data: function data() {
    return {
      vmOverlay: null,
      overlayOpened: false,
      showOverlay: false,
      chnageTranslate: baseFn.ua.isWechatMiniProgram() && baseFn.ua.isIPhoneX()
    };
  },
  computed: {
    transitionName: function transitionName() {
      var transitionName = 'popup-slide-';

      if (this.position) {
        transitionName += this.position;
      } else {
        transitionName = 'popup-fade';
      }

      return transitionName;
    }
  },
  watch: {
    value: function value(val) {
      var type = val ? 'open' : 'close';
      this[type]();
      this.$emit(type);
    }
  },
  mounted: function mounted() {
    if (this.value) {
      this.open();
    }
  },
  activated: function activated() {
    if (this.value) {
      this.open();
    }
  },
  beforeDestroy: function beforeDestroy() {
    this.close();
  },
  deactivated: function deactivated() {
    this.close();
  },
  methods: {
    closePopup: function closePopup() {
      if (this.closeOnClickOverlay) {
        this.$emit('input', false);
      }

      this.$emit('click-overlay');
    },
    open: function open() {
      if (this.overlay) {
        this.showOverlay = true;
      }
    },
    close: function close() {
      this.showOverlay = false;
      this.$emit('input', false);
      this.unBindFixed();
    },
    unBindFixed: function unBindFixed() {
      if (this.fixed) {
        if (window.currentRouterView) {
          window.currentRouterView.style.overflow = 'scroll !important';
        } else {
          var body = document.body;
          body.style.position = '';
          var top = body.style.top;
          document.body.scrollTop = document.documentElement.scrollTop = -parseInt(top);
          body.style.top = '';
        }
      }
    }
  }
};

var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "popup__root"
  }, [_c('transition', {
    attrs: {
      "name": _vm.transitionName,
      "duration": _vm.duration
    },
    on: {
      "afterEnter": function afterEnter($event) {
        return _vm.$emit('opened');
      },
      "afterLeave": function afterLeave($event) {
        return _vm.$emit('closed');
      }
    }
  }, [_vm.value ? _c('div', {
    directives: [{
      name: "fixed",
      rawName: "v-fixed",
      value: _vm.fixed,
      expression: "fixed"
    }],
    staticClass: "popup__slot--container components__popup",
    "class": 'popup__' + (_vm.position || 'center'),
    staticStyle: {
      "z-index": "4000"
    },
    style: _vm.slotStyle,
    on: {
      "click": function click($event) {
        return _vm.$emit('slotClick');
      }
    }
  }, [_c('div', {
    style: _vm.slotStyle
  }, [_c('div', {
    staticClass: "translate__wapper",
    "class": [_vm.wapperClass, {
      'translate__wapper_ipx': _vm.chnageTranslate
    }],
    style: Object.assign({}, {
      transition: _vm.duration / 1000 + 's'
    }, _vm.slotStyle)
  }, [_vm._t("default")], 2)])]) : _vm._e()]), _vm._v(" "), _c('overlay', {
    staticClass: "overlay",
    attrs: {
      "value": _vm.showOverlay,
      "overlay-style": _vm.overlayStyle
    },
    on: {
      "click": _vm.closePopup
    }
  })], 1);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-0153090e_0", {
    source: ".popup__slot--container[data-v-0153090e]{position:fixed;max-height:100%;pointer-events:none}.popup__slot--container .translate__wapper[data-v-0153090e]{transition:.3s;max-height:100%;pointer-events:auto}.popup__center[data-v-0153090e]{top:50%;left:50%}.popup__center .translate__wapper[data-v-0153090e]{opacity:1;transform:translate(-50%,-50%)}.popup__center .translate__wapper_ipx[data-v-0153090e]{transform:translate(-50%,calc(-50% - .5rem))}.popup__top[data-v-0153090e]{width:100%;top:0;right:auto;bottom:auto;left:0}.popup__bottom[data-v-0153090e]{width:100%;top:auto;bottom:0;right:auto;left:0}.popup__right[data-v-0153090e]{top:50%;right:0;bottom:auto;left:auto}.popup__right .translate__wapper[data-v-0153090e]{transform:translate(0,-50%)}.popup__left[data-v-0153090e]{top:50%;left:0;bottom:auto;right:auto}.popup__left .translate__wapper[data-v-0153090e]{transform:translate(0,-50%)}.popup-slide-top-enter .translate__wapper[data-v-0153090e],.popup-slide-top-leave-active .translate__wapper[data-v-0153090e]{transform:translate(0,-100%)}.popup-slide-right-enter .translate__wapper[data-v-0153090e],.popup-slide-right-leave-active .translate__wapper[data-v-0153090e]{transform:translate(100%,-50%)}.popup-slide-bottom-enter .translate__wapper[data-v-0153090e],.popup-slide-bottom-leave-active .translate__wapper[data-v-0153090e]{transform:translate(0,100%)}.popup-slide-left-enter .translate__wapper[data-v-0153090e],.popup-slide-left-leave-active .translate__wapper[data-v-0153090e]{transform:translate(-100%,-50%)}.popup-fade-enter .translate__wapper[data-v-0153090e],.popup-fade-leave-active .translate__wapper[data-v-0153090e]{opacity:0;transform:translate(-50%,-50%) scale(.8,.8)}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-0153090e";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = /*#__PURE__*/vueRuntimeHelpers.normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, vueRuntimeHelpers.createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

exports.default = __vue_component__;
