// @wanwu/base-vue-occupying/lib/index.es.js
import 'core-js/modules/es.function.name';
import 'core-js/modules/es.number.constructor';
import { normalizeComponent, createInjector } from '@wanwu/vue-runtime-helpers';

//
//
//
//
//
//
//
//
var script = {
  name: 'Occupying',
  props: {
    height: {
      "default": 20,
      type: [Number, String]
    },
    bgColor: {
      "default": '',
      type: String
    }
  },
  computed: {
    remHeight: function remHeight() {
      return this.height / 100 + 'rem';
    }
  }
};

/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: "component__Occupying Occupying",
    style: {
      height: _vm.remHeight,
      backgroundColor: _vm.bgColor
    }
  }, [_vm._t("default")], 2);
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-27f4ac77_0", {
    source: ".component__Occupying[data-v-27f4ac77]{width:100%}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-27f4ac77";
/* module identifier */

var __vue_module_identifier__ = undefined;
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject SSR */

/* style inject shadow dom */

var __vue_component__ = /*#__PURE__*/normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, createInjector, undefined, undefined);

__vue_component__.install = function (Vue) {
  Vue.component(__vue_component__.name, __vue_component__);
};

export default __vue_component__;
